# info --------------------------------------------------------------------
# purpose: modify S288C reference genome to improve mapping of W303-derived reads; add strain-specific mutations
# author: Robert Gnuegge (robert.gnuegge@gmail.com)
# date: 01/28/22
# version: 1.0

# preamble ----------------------------------------------------------------

# set wd to this file's location
wd_path <- dirname(rstudioapi::getActiveDocumentContext()$path)
setwd(wd_path)

# libraries
library(GenomicRanges)
library(Biostrings)
library(rtracklayer)


# load S288C reference genome ---------------------------------------------
# sequence downloaded from http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/ on 05/11/20

genome <- import(con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/S288C_reference_sequence_R64-2-1_20150113.fasta")

# change names
names(genome)
names(genome) <- c(paste0("chr", as.roman(1:16)), "chrM")
names(genome)


# add 2 micron plasmid ----------------------------------------------------
# sequence downloaded from https://www.ncbi.nlm.nih.gov/nuccore/CM001822.1 on 05/11/20

micron <- import(con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/2micron_W303.fasta")

# the 2 micron plasmid is circular
# to support read mapping near its 3' end, add a sequence stretch from its 5' end
read_length <- 75
micron$micron <- c(micron$micron, micron$micron[1:(read_length - 1)])

# add to genome
genome <- DNAStringSet(x = c(as.character(genome), as.character(micron)))
seqinfo(genome)

# inject W303 SNPs --------------------------------------------------------
# SNPs downloaded from https://www.g3journal.org/highwire/filestream/485032/field_highwire_adjunct_files/0/FileS1.xlsx on 05/11/20
# (Matheson et al., 2017)

SNPs <- read.csv(file = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/W303_SNPs.csv", header = TRUE, stringsAsFactors = FALSE)

# change chrmt to chrM
SNPs$Chromosome[SNPs$Chromosome == "chrmt"] <- "chrM"

# check that the S288C nts in the SNPs file are consistent with the genome sequence
all(sapply(X = names(genome)[1:17], FUN = function(c) {as.character(genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]]) == paste0(SNPs$Base.in.S288c[SNPs$Chromosome == c], collapse = "")}))

# substitute with W303 SNPs
for(c in c(paste0("chr", as.roman(1:16)), "chrM")){
  genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]] <- DNAString(paste0(SNPs$Base.in.W303[SNPs$Chromosome == c], collapse = ""))
}

# check that SNPs have been injected
all(sapply(X = names(genome)[1:17], FUN = function(c) {as.character(genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]]) == paste0(SNPs$Base.in.W303[SNPs$Chromosome == c], collapse = "")}))


# change MATalpha to MATa -------------------------------------------------
# the strains used for S1-seq library preparation are all MATa

# correct HMRa1 stk mutation (see https://www.yeastgenome.org/locus/S000029699#paragraph for details)
matchPattern(pattern = DNAString("CGCAACAGTAAAATTTTATAA"), subject = genome[["chrIII"]])
matchPattern(pattern = DNAString("CGCAACAGTATAATTTTATAA"), subject = genome[["chrIII"]])
# already corrected during the W303 SNP injection above

# define MATalpha Z region (coordinates from https://www.yeastgenome.org/locus/S000029699)
Z1 <- genome[["chrIII"]][200851:201089]

# find Z regions in HML, MAT, and HMR
all.Z1 <- matchPattern(pattern = Z1, subject = genome[["chrIII"]], max.mismatch = 1)

# turns out that there is a 1 bp difference between HML/MAT and HMR (C->T)
mismatch(pattern = Z1, x = all.Z1)

# define MATalpha X region
X <- genome[["chrIII"]][199401:200103] 

# find X regions in HML, MAT, and HMR
all.X <- matchPattern(pattern = X, subject = genome[["chrIII"]], max.mismatch = 2, with.indels = TRUE)

# MATalpha has a 2 bp deletion at position 199512:199513 (in a poly-TA stretch)
mismatch(pattern = X, x = all.X)  # gives mismatches for HML and HMR at relative position > 111

# let's only use the first 111 bp of the MATalpha X
X <- genome[["chrIII"]][199401:199511] 

# find X regions in HML, MAT, and HMR
all.X <- matchPattern(pattern = X, subject = genome[["chrIII"]], max.mismatch = 0)

# MATalpha has a 2 bp deletion at position 199512:199513 (in a poly-TA stretch)
mismatch(pattern = X, x = all.X)  # gives mismatches for HML and HMR at relative position > 111

# check size difference of regions
end(all.Z1[2]) - start(all.X[2]) - ( end(all.Z1[3]) - start(all.X[3]) )
# MATalpha X-Z region is 103 bp longer than MATa X-Z region

# replace MATalpha with HMR and add 103 Ns downstream of TAF1 gene (ca. 5 kb downstream of HO cut site)
TAF1.end <- 206260  # upstream of YCR043C, which is downstream of TAF1 (https://www.yeastgenome.org/locus/YCR043C)
MATalpha <- genome[["chrIII"]][start(all.X[2]):TAF1.end]

HMR <- genome[["chrIII"]][start(all.X[3]):end(all.Z1[3])]
TAF1 <- genome[["chrIII"]][(end(all.Z1[2]) + 1):TAF1.end]
Ns <- DNAString(x = paste0(rep("N", 103), collapse = ""))
replacement <- c(HMR, TAF1, Ns)

# sanity check
length(MATalpha) == length(replacement)

# replace
genome[["chrIII"]][start(all.X[2]):TAF1.end] <- replacement

# check if HMR sequence was copied to MAT
matchPattern(pattern = HMR, subject = genome[["chrIII"]])

# check HO cut sites
matchPattern(pattern = DNAString(x = "GCAACAGTATAATTTTATAAACCCT"), subject = genome[["chrIII"]])
# the middle hit is MAT


# HML deletion ------------------------------------------------------------
# the HML deletion was generated by C. Zierhut (https://www.ncbi.nlm.nih.gov/pubmed/18511906 and personal communication)
# HML was replaced by URA3, which was then in turn replaced by a PCR fragment derived from pRS306 (ori) using the following primers
OCZ330 <- DNAString(x = "gacaagtagcgcagttattttcttattttcatCATCAGAGCAGATTGTACTGAGAGTGCAtaattgcgttgcgctcactg")
OCZ331 <- reverseComplement(x = DNAString(x = "CAAATTTACTAGCTcgagagttagttaatccttactaagtgaagaaaagcaactatagtcTCAAGAACTCTGTAGCACCG"))

# find replaced genome region
HML.start <- start(matchPattern(pattern = OCZ330[1:32], subject = genome[["chrIII"]]))
HML.end <- end(matchPattern(pattern = OCZ331[21:80], subject = genome[["chrIII"]]))
HML <- genome[["chrIII"]][HML.start:HML.end]

# find amplified pRS306 region
# pRS306 sequence downloaded from https://www.ncbi.nlm.nih.gov/nuccore/U03438.1/ on 05/11/20
pRS306 <- import(con = "~/Research/Resources/pRS306.fasta")[[1]]
ori.start <- start(matchPattern(pattern = OCZ330[61:80], subject = pRS306))
ori.end <- end(matchPattern(pattern = OCZ331[1:20], subject = pRS306))
ori <- pRS306[ori.start:ori.end]
# expand with primer overhangs
replacement <- c(OCZ330[1:60], ori, OCZ331[21:80])

# length of replacement is 3495 bp shorter than replaced region
length(HML) - length(replacement)
# let's fill up with Ns in the center of the replacement
replacement <- c(replacement[1:(round(0.5 * length(replacement)) - 1)],
                 DNAString(x = paste0(rep("N", 3495), collapse = "")),
                 replacement[round(0.5 * length(replacement)):length(replacement)])

# sanity check
length(HML) == length(replacement)

# replace
genome[["chrIII"]][HML.start:HML.end] <- replacement

# check if ori parts were inserted into genome
matchPattern(pattern = ori[1:floor(0.5 * length(ori) - 1)], subject = genome[["chrIII"]])
matchPattern(pattern = ori[ceiling(0.5 * length(ori) + 1):length(ori)], subject = genome[["chrIII"]])
matchPattern(pattern = HML, subject = genome[["chrIII"]])

# check HO cut sites
matchPattern(pattern = DNAString(x = "GCAACAGTATAATTTTATAAACCCT"), subject = genome[["chrIII"]])
# HOcs at MATa and HMR retained

# mask endogenous GAL1 promoter and T_CYC1 promoter -----------------------
# (they are in the reporter construct)

T_CYC1 <- DNAString(x = "ACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTG")
matchPattern(pattern = T_CYC1, subject = genome[["chrX"]], max.mismatch = 1)
genome[["chrX"]][matchPattern(pattern = T_CYC1, subject = genome[["chrX"]], max.mismatch = 1)] <- DNAString(x = paste0(rep("N", 251), collapse = ""))
matchPattern(pattern = T_CYC1, subject = genome[["chrX"]], max.mismatch = 1)

P_GAL1 <- DNAString(x = "AGTACGGATTAGAAGCCGCCGAGCGGGCGACAGCCCTCCGACGGAAGACTCTCCTCCGTGCGTCCTCGTCTTCACCGGTCGCGTTCCTGAAACGCAGATGTGCCTCGCGCCGCACTGCTCCGAACAATAAAGATTCTACAATACTAGCTTTTATGGTTATGAAGAGGAAAAATTGGCAGTAACCTGGCCCCACAAACCTTCAAATTAACGAATCAAATTAACAACCATAGGATGATAATGCGATTAGTTTTTTAGCCTTATTTCTGGGGTAATTAATCAGCGAAGCGATGATTTTTGATCTATTAACAGATATATAAATGGAAAAGCTGCATAACCACTTTAACTAATACTTTCAACATTTTCAGTTTGTATTACTTCTTATTCAAATGTCATAAAAGTATCAACAAAAAATTGTTAATATACCTCTATACTTTAACGTCAAGGAGAAAAAACTATA")
matchPattern(pattern = P_GAL1, subject = genome[["chrII"]], max.mismatch = 1)
genome[["chrII"]][matchPattern(pattern = P_GAL1, subject = genome[["chrII"]], max.mismatch = 1)] <- DNAString(x = paste0(rep("N", 457), collapse = ""))
matchPattern(pattern = P_GAL1, subject = genome[["chrII"]], max.mismatch = 1)


# add HMR-HOcs modification (LSY4753-3B) ==================================

# Seems there is a length limit of DNAStrings of 4077 nt.
# We need to split our replacement string into two parts.
pRG726_up <- DNAString(x = "CCTACGATTACTATGTACTGGAAGGTGAGTCTTGAATTCATCCTCAGAGATTGTTCTAAAGACCAACAGAAATCCATATAGATTACCGTAGTTACTCTGATGGAATTTCAGAGTAAGACCCTTGTCAACTTCATCACTTTAAGTTGAGATGTAGAAAGACACTTTTGTAAGGTCTCCAACAAGTATATTTGATAAATAGACTGTTGAGCTACAAAAACTCAATGTGAAGTAAAAGACCTCATAAATGATCGATTATCTGTTTGGCAGGTACTAATAAAACTGCATTGGTGTCTTTTTTCACTTACGTTAAAAAAAACTTTTTGCAATGATTTAAGCGTGCGTGAAGATAACACTACAATCCATTTTAAAGCAACATCCACATTGAGTGTATACACCACAAAGGTTTTTTCAGGGCGTTTTTCTCGCCACTTTATGTTGACCAAAATTATTAATGGAACTTACAACGTTTCCAAAAGTTAGTTAAATACATACGTCTATTTACTAAGCAAGAAATATATCATGACAAGCCCAAATATTATATTGTTATGTTTACAAAAAAAAAATGGCTATATACATCAAGTCTGGAGGCTTTTTATAACAAGCAAGTGGGGTAACTTAGACATAAGATTGACTTCTTTGAATTCAACAAAAATACATACTTTTGATGATTTCAATGGTAGAAGCAGTAGTAACTATGCCGATGAAATAAAATCCTGATATCATCTATATAGTAGTACTCACGTTACTAGTATATCATCTATAAGATGTAGGAAAATGTCATAAGAAATGAGAAACAGTCATTAAATTTAATGGAAACGAAATGCAAGGATTGGTGATGAGATAAGATAATGAAACATAGAAAACGAAAGGAGAAATAACAGTAGTATTATTATTATTATTATTATGTAGAAATATTGACTCTCTCTTGTAGATTCATATATCTTCAAGGGGAACTTCTTGTACATTTTGTATACCTATATTATAGCGTTTAACAACAATGGAATCCCAACAATTACCAAAATTTTCACACATTCCTCATATAAAATCTAAGCAATGACTAGAGAACTATCGCTAAATTATATAGAGCAAGGAAACCGAATAATTACATGTGTCAGATAAGAAAAAAATAAAGAAAACAACGCGTCATGAAAAAGAGTTACTGAACCTTCAGATCCTACTTATTGTAATGCTTCGCGACATCCAATCCATTTAATAATCAATTTAAAACTAGAGTTGGTAGAGTTCCTTGTTGAACGTGATAACCCAAAAGCATAATACGAGTAATGTTTCAGTATTGCTATTATATGTTTACACAAGGAAAACATATAATAACAAACCTCTAATCCGGTAGTACTTAAGAAACTATAGTTTCTATGTACAAAAAGGTAACTATGTAATTCTTACATTTACATAACATATAGAAGGGTCCAATAAACTTACTAAACTTACTACCTTGTTGTATATAGGCTAGATCGTAATCCACTACGTCAACATAAAAAAAAAACTTAAGAAGTTTGAATTTTATGTACAAACAGATTGTTAAAATATAATATAAGATTATGGAAACGAACTTGGTCTAAAAAAAATTTAAAGTTTTATAAAATCCTCGAACTATCGCTGTTATACATGATGTCCCCAAAGCGTGTACAAATAATTTTGTAGTATTGTATCGGTAATATCATACAAGAAGAAAATAGTTGCAAATCTCTAACCCACTAGTACTTAAAAAAACTGTAGTTTCAGTGCAAAAAAGTTTTAACATTACGTATCTTGTACCCTTTTTATTGCATATAGAAAGGTCAAATAATCCTTCACATCATGAAATATAAGCTAAATCGCATTTCTTTTCGTCCACATTTGCAAACAAAACTTTTCAATAATAATTTTATAAATAGTATCAATATATATATATATATATTTATTTGTTTACTTTTTCTATCAGTGTTTTCAATTTTTTATTAAACAATGTTTGATTTTTTAAATCGCAATTTAATACCTAAATATAAAAAATGTTATTATATTGCAAAAACCCATCAACCTTGAAAAAAAGTAGAAACGTTTTATTTAATTCTATCAATACATCATAAAATACGAACGATCCCCGTCCAAGTTATGAGCTTAATCTTCCATAAAAATATTTGAAAGCAATAGATCATGTACTAAACTAAAATCAGGGAAATTAAGACTCCTTTTGAAGTAATACCTATTACTTACTAATACGTTTGAGAATAAGCGCAGGTACTCCTGGTTTTTGTTAAAATTACAAATTTATACTTAGCATTACGAAGATTCTCGATTCCGAAAAACAAAAATTTTATCGTCATATACAAATCTAGAAATTACCAGAGCTATCGCGGCCGCAGTACGGATTAGAAGCCGCCGAGCGGGCGACAGCCCTCCGACGGAAGACTCTCCTCCGTGCGTCCTCGTCTTCACCGGTCGCGTTCCTGAAACGCAGATGTGCCTCGCGCCGCACTGCTCCGAACAATAAAGATTCTACAATACTAGCTTTTATGGTTATGAAGAGGAAAAATTGGCAGTAACCTGGCCCCACAAACCTTCAAATTAACGAATCAAATTAACAACCATAGGATGATAATGCGATTAGTTTTTTAGCCTTATTTCTGGGGTAATTAATCAGCGAAGCGATGATTTTTGATCTATTAACAGATATATAAATGGAAAAGCTGCATAACCACTTTAACTAATACTTTCAACATTTTCAGTTTGTATTACTTCTTATTCAAATGTCATAAAAGTATCAACAAAAAATTGTTAATATACCTCTATACTTTAACGTCAAGGAGAAAAAACTATAACTAGTATGAGATCTAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAATAA")
pRG726_down <- DNAString(x = "CTCGAGACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGATCCatttggccttatagagtgtggtcgtggcggaggttgtttatctttcgagtactgaatgttgtcagtatagctatcctatttgaaactccccatcgtcttgctcttgttcccaatgtttgtttataTACTCATAtTTCTATGTGTTTATACAATTGCTATTGTTTATATAATGTAGTGACATTTTCTCTTAATCTTATACTAATTTCTATGACATTTATATAAGAAGAGACTTATGATCAACATAATTTTGCAAACTTTGAGAGAAATATGTCTTTCTACTGCGATAAAGTTATTATTTAGATTACATGTCACCAACATTTTCGTATATGGCGATATAATTTATCATGTTTTGGTATGATAATTTAATTTTTAAAAAAACAAATTTAATTGACCTCATTAATTAATATTTATTAATACCTTTAAATGTTGAGGTAAATAGCTATTTTCTCTCTTCTTTTCCTTTAGTTGGAATTTGCACAAGAAAATGTTTTTCCACACACTTTAGCGTTTTTTCCTAAAGTGTTGCCAGCGATAAACGGCAGCCGCCGAAGTCAAGACGGCGCAGGGTTAAATAGCGTGGATCTTCGCCCTGTGCGCGGCGGTTGAGTAGCTCTTCTTCCAGCACGTTTTCGCCGATAATACCGAGATCGACCACGCCATCCATTACCAGACCCGGAATGTCATCATCACGCACGCGCAGGATATCAATCGGCATGTTTTCCGCCATCGCAATCAGGCGCTGAGTGTGTAAATTAATTTTTATGCCGCAGCGGGCCAGCAATTCTCGTGAATCATCGCTTAAACGGCCTGATTTCTGAATAGCTATGCGTAAGCGGGTGTTGTCTAATCTGCAGATTACTTTCCTAAATTTGCATATAGAATTGTCAAGCGCAAATCCGACGTCGATTCCGCGGCGGATGGGTCATTCTAGGTCATTCTACCAATTTTATTTGAGACCAGGTTTATTCAACCGGTAACATAGAAATATTCATACAATTAAGCTTCTATGGCCAAGTTGGTAAGGCGCCACACTAGTAATGTGGAGATCATCGGTTCAAATCCGATTGGAAGCATTTTTTATCACGTTATTCGGTGACACCCAGGTTGCCGCCGCGTTCGCGTCCATCGTCATCTGAAAAATAATGAATATTTTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTTTACTTTTGTATATTAAATTTACCGTTATTCGGAGATCTCTTACGGCTTATGATTTTCTTTACATTCCAGGCCGCCTTTTGATGACATTCAGTGCGTCACGACGCGAATAGGGCTTTGAAAGCCTTAATGTGGAGTAAATAAATATATGAATGCAATAAGTAAACATAAGACGAGTTCTTCTATATCCGGTGTACCTAATAATATAGATTTTGCCAACAATGGAATCTCATTACCCATCCCAAGATAAAATATTGATCAAGGCATGGGCCACCAACATAGGAATACGCAAAATAGGATTGATATTTTTACATATTATATTATCTTTCTTCTTCTGTCTCAAATATAGCTCTTCTCCTATTTAAAAAATTATCAATTCTTGAATCTCAACTTCCATTAAATTTGATGATTGTTTCCCAATATCGATGTCATCTTTTTACACTGTAAATTATAATTTACTAGGAACATGAATACTGATCAATGGATGATAGTTGATTCTTATTGCAATAAAAAGTACTTTCTAAGACGCTAGGACTTCTAAACACAGATATTATGAAAATGCCAAAAAATACAGAATATTTTCCTTCTTTTCACTATAACAAGTAGTAAGAAATTAAGCAATATGGTAATTTGAGCTTATCAGATAAAAGATCCTGGTAGTCTTGCAAAGGTGTGAGTACTTTTGATGAAACTAAGTTTCAGCGCAGCTATTTCAATTTTGGAACCTTTAAAGAAAGTTTAATCGAAACATTGAACGATCCAAAGCACAACTTTGTAGCGGGAAGTCACAAACAAATATAAAAAGTGTTAATAAACTGCCATCCTATGTTAGAATTGTGAAACTTTTTGGACTAATAAAAGGTTTTCAACATGGTGTTCCAAAGCACTTATTTCTAGGTTAGCTATGGTAACGAGTTTTAACGTAATCAAAACAATATACAATTCAAGATCTGTATCATTTACACTAACACACCCCATTTTAATAGAGCTGAGTACTTCCTACATTTCTTTTGCATATGATATCGATACTGCATCATTTCTGTAGTCGTAAAATTTCCTTCTCATTAAGACACCGAAAATAGCTAAGGATTATAACAAAACAAACAAATGCTCAAACAAAACTTTTAGTAAATCTTTTGAAATGTAAAAGTTACTTAGAATAACTAAAAAAAAAATATTACGCTATACCATATATTTCCATTCCCAATTTTATCGTTTTATAATAATATTCAGTTTCAAGAACAATGAATATATTTCAAACATTACATGCACTCTTGGGAGTAAGCTGCATCATAATAAATTAAAATTAACGCTTTAGAAAAAGAAATTTTGAGGTTAAAGGATGGAGG")

pRG726 <- c(pRG726_up, pRG726_down)

# find corresponding coordinates in genome
start <- start(matchPattern(pattern = pRG726[1:50], subject = genome[["chrIII"]]))
end <- end(matchPattern(pattern = pRG726[(length(pRG726)-50):length(pRG726)], subject = genome[["chrIII"]]))

# the replacement seems to be a 2 nt shorter than the genomic sequence
end - start
length(pRG726)

# Let's replace with a duplication of first two nts at the replacement site (about 5 kb upstream of DSB sites).
genome[["chrIII"]][(start + 2):end] <- pRG726

# are +/- 1 kb around DSB site present?
HOcs_flanking <- DNAString(x = "GACATTTTCTCTTAATCTTATACTAATTTCTATGACATTTATATAAGAAGAGACTTATGATCAACATAATTTTGCAAACTTTGAGAGAAATATGTCTTTCTACTGCGATAAAGTTATTATTTAGATTACATGTCACCAACATTTTCGTATATGGCGATATAATTTATCATGTTTTGGTATGATAATTTAATTTTTAAAAAAACAAATTTAATTGACCTCATTAATTAATATTTATTAATACCTTTAAATGTTGAGGTAAATAGCTATTTTCTCTCTTCTTTTCCTTTAGTTGGAATTTGCACAAGAAAATGTTTTTCCACACACTTTAGCGTTTTTTCCTAAAGTGTTGCCAGCGATAAACGGCAGCCGCCGAAGTCAAGACGGCGCAGGGTTAAATAGCGTGGATCTTCGCCCTGTGCGCGGCGGTTGAGTAGCTCTTCTTCCAGCACGTTTTCGCCGATAATACCGAGATCGACCACGCCATCCATTACCAGACCCGGAATGTCATCATCACGCACGCGCAGGATATCAATCGGCATGTTTTCCGCCATCGCAATCAGGCGCTGAGTGTGTAAATTAATTTTTATGCCGCAGCGGGCCAGCAATTCTCGTGAATCATCGCTTAAACGGCCTGATTTCTGAATAGCTATGCGTAAGCGGGTGTTGTCTAATCTGCAGATTACTTTCCTAAATTTGCATATAGAATTGTCAAGCGCAAATCCGACGTCGATTCCGCGGCGGATGGGTCATTCTAGGTCATTCTACCAATTTTATTTGAGACCAGGTTTATTCAACCGGTAACATAGAAATATTCATACAATTAAGCTTCTATGGCCAAGTTGGTAAGGCGCCACACTAGTAATGTGGAGATCATCGGTTCAAATCCGATTGGAAGCATTTTTTATCACGTTATTCGGTGACACCCAGGTTGCCGCCGCGTTCGCGTCCATCGTCATCTGAAAAATAATGAATATTTTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTTTACTTTTGTATATTAAATTTACCGTTATTCGGAGATCTCTTACGGCTTATGATTTTCTTTACATTCCAGGCCGCCTTTTGATGACATTCAGTGCGTCACGACGCGAATAGGGCTTTGAAAGCCTTAATGTGGAGTAAATAAATATATGAATGCAATAAGTAAACATAAGACGAGTTCTTCTATATCCGGTGTACCTAATAATATAGATTTTGCCAACAATGGAATCTCATTACCCATCCCAAGATAAAATATTGATCAAGGCATGGGCCACCAACATAGGAATACGCAAAATAGGATTGATATTTTTACATATTATATTATCTTTCTTCTTCTGTCTCAAATATAGCTCTTCTCCTATTTAAAAAATTATCAATTCTTGAATCTCAACTTCCATTAAATTTGATGATTGTTTCCCAATATCGATGTCATCTTTTTACACTGTAAATTATAATTTACTAGGAACATGAATACTGATCAATGGATGATAGTTGATTCTTATTGCAATAAAAAGTACTTTCTAAGACGCTAGGACTTCTAAACACAGATATTATGAAAATGCCAAAAAATACAGAATATTTTCCTTCTTTTCACTATAACAAGTAGTAAGAAATTAAGCAATATGGTAATTTGAGCTTATCAGATAAAAGATCCTGGTAGTCTTGCAAAGGTGTGAGTACTTTTGATGAAACTAAGTTTCAGCGCAGCTATTTCAATTTTGGAACCTTTAAAGAAAGTTTAATCGAAACATTGAACGATCCAAAGCACAACTTTGTAGCGGGAAGTCACAAACAAATATAAAAAGTGTTAATAAACTGCCATCCTATGTTAGAATTGTGAAACTTTTTGGACTAATAAAAGGTTTTCAACATGGTGTTCCAAAGCACTTATTTCTAGGTTAGCTATGGTAACGAGTTTTAACGTAATCAAAACAATATACAATTCAAGATCTGTATCATTTACACTAACACACCCCAT")
matchPattern(pattern = HOcs_flanking, subject = genome[["chrIII"]])

# are there two HOcs now on chrIII?
matchPattern(pattern = "ttCTTTTaGTTTCAGCTTTCCGCAACAGTATAATTTTATAAAcCCTGGtt", subject = genome[["chrIII"]])

# mask MATa/HMR duplicated sequence from MATa
duplication <- DNAString(x = "CATTTGGCCTTATAGAGTGTGGTCGTGGCGGAGGTTGTTTATCTTTCGAGTACTGAATGTTGTCAGTATAGCTATCCTATTTGAAACTCCCCATCGTCTTGCTCTTGTTCCCAATGTTTGTTTATATACTCATAT")
matchPattern(pattern = duplication, subject = genome[["chrIII"]])
genome[["chrIII"]][matchPattern(pattern = duplication, subject = genome[["chrIII"]])[1]] <- DNAString(x = paste0(rep("N", length(duplication)), collapse = ""))
matchPattern(pattern = duplication, subject = genome[["chrIII"]])

# export modified reference genome to fast file
export(object = genome, con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Data/S_cerevisiae_reference_genomes/S288C_R64-2-1_W303_SNPs_MATa_hml_hmr_LSY4753-3B.fasta")


# add hmr-HOcs modification (LSY4754-3A) ==================================

# Seems there is a length limit of DNAStrings of 4077 nt.
# We need to split our replacement string into two parts.
pRG727_up <- DNAString(x = "CCTACGATTACTATGTACTGGAAGGTGAGTCTTGAATTCATCCTCAGAGATTGTTCTAAAGACCAACAGAAATCCATATAGATTACCGTAGTTACTCTGATGGAATTTCAGAGTAAGACCCTTGTCAACTTCATCACTTTAAGTTGAGATGTAGAAAGACACTTTTGTAAGGTCTCCAACAAGTATATTTGATAAATAGACTGTTGAGCTACAAAAACTCAATGTGAAGTAAAAGACCTCATAAATGATCGATTATCTGTTTGGCAGGTACTAATAAAACTGCATTGGTGTCTTTTTTCACTTACGTTAAAAAAAACTTTTTGCAATGATTTAAGCGTGCGTGAAGATAACACTACAATCCATTTTAAAGCAACATCCACATTGAGTGTATACACCACAAAGGTTTTTTCAGGGCGTTTTTCTCGCCACTTTATGTTGACCAAAATTATTAATGGAACTTACAACGTTTCCAAAAGTTAGTTAAATACATACGTCTATTTACTAAGCAAGAAATATATCATGACAAGCCCAAATATTATATTGTTATGTTTACAAAAAAAAAATGGCTATATACATCAAGTCTGGAGGCTTTTTATAACAAGCAAGTGGGGTAACTTAGACATAAGATTGACTTCTTTGAATTCAACAAAAATACATACTTTTGATGATTTCAATGGTAGAAGCAGTAGTAACTATGCCGATGAAATAAAATCCTGATATCATCTATATAGTAGTACTCACGTTACTAGTATATCATCTATAAGATGTAGGAAAATGTCATAAGAAATGAGAAACAGTCATTAAATTTAATGGAAACGAAATGCAAGGATTGGTGATGAGATAAGATAATGAAACATAGAAAACGAAAGGAGAAATAACAGTAGTATTATTATTATTATTATTATGTAGAAATATTGACTCTCTCTTGTAGATTCATATATCTTCAAGGGGAACTTCTTGTACATTTTGTATACCTATATTATAGCGTTTAACAACAATGGAATCCCAACAATTACCAAAATTTTCACACATTCCTCATATAAAATCTAAGCAATGACTAGAGAACTATCGCTAAATTATATAGAGCAAGGAAACCGAATAATTACATGTGTCAGATAAGAAAAAAATAAAGAAAACAACGCGTCATGAAAAAGAGTTACTGAACCTTCAGATCCTACTTATTGTAATGCTTCGCGACATCCAATCCATTTAATAATCAATTTAAAACTAGAGTTGGTAGAGTTCCTTGTTGAACGTGATAACCCAAAAGCATAATACGAGTAATGTTTCAGTATTGCTATTATATGTTTACACAAGGAAAACATATAATAACAAACCTCTAATCCGGTAGTACTTAAGAAACTATAGTTTCTATGTACAAAAAGGTAACTATGTAATTCTTACATTTACATAACATATAGAAGGGTCCAATAAACTTACTAAACTTACTACCTTGTTGTATATAGGCTAGATCGTAATCCACTACGTCAACATAAAAAAAAAACTTAAGAAGTTTGAATTTTATGTACAAACAGATTGTTAAAATATAATATAAGATTATGGAAACGAACTTGGTCTAAAAAAAATTTAAAGTTTTATAAAATCCTCGAACTATCGCTGTTATACATGATGTCCCCAAAGCGTGTACAAATAATTTTGTAGTATTGTATCGGTAATATCATACAAGAAGAAAATAGTTGCAAATCTCTAACCCACTAGTACTTAAAAAAACTGTAGTTTCAGTGCAAAAAAGTTTTAACATTACGTATCTTGTACCCTTTTTATTGCATATAGAAAGGTCAAATAATCCTTCACATCATGAAATATAAGCTAAATCGCATTTCTTTTCGTCCACATTTGCAAACAAAACTTTTCAATAATAATTTTATAAATAGTATCAATATATATATATATATATTTATTTGTTTACTTTTTCTATCAGTGTTTTCAATTTTTTATTAAACAATGTTTGATCTCCATCATCTTCTCGATCGGCAGTACCAGAATCGAGCTGGCGCCAAGCGCTTTCAGTTTCTCCATGGTTTCCCAGAACAACGTTTCGCTGCTGACCATGTGCATCGCCACGCGCTGTTAACGATCCCCGTCCAAGTTATGAGCTTAATCTTCCATAAAAATATTTGAAAGCAATAGATCATGTACTAAACTAAAATCAGGGAAATTAAGACTCCTTTTGAAGTAATACCTATTACTTACTAATACGTTTGAGAATAAGCGCAGGTACTCCTGGTTTTTGTTAAAATTACAAATTTATACTTAGCATTACGAAGATTCTCGATTCCGAAAAACAAAAATTTTATCGTCATATACAAATCTAGAAATTACCAGAGCTATCGCGGCCGCAGTACGGATTAGAAGCCGCCGAGCGGGCGACAGCCCTCCGACGGAAGACTCTCCTCCGTGCGTCCTCGTCTTCACCGGTCGCGTTCCTGAAACGCAGATGTGCCTCGCGCCGCACTGCTCCGAACAATAAAGATTCTACAATACTAGCTTTTATGGTTATGAAGAGGAAAAATTGGCAGTAACCTGGCCCCACAAACCTTCAAATTAACGAATCAAATTAACAACCATAGGATGATAATGCGATTAGTTTTTTAGCCTTATTTCTGGGGTAATTAATCAGCGAAGCGATGATTTTTGATCTATTAACAGATATATAAATGGAAAAGCTGCATAACCACTTTAACTAATACTTTCAACATTTTCAGTTTGTATTACTTCTTATTCAAATGTCATAAAAGTATCAACAAAAAATTGTTAATATACCTCTATACTTTAACGTCAAGGAGAAAAAACTATA")
pRG727_down <- DNAString(x = "ACTAGTATGAGATCTAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAATAACTCGAGACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGATCCatttggccttatagagtgtggtcgtggcggaggttgtttatctttcgagtactgaatgttgtcagtatagctatcctatttgaaactccccatcgtcttgctcttgttcccaatgtttgtttataTACTCATATTTCTATGTGTTTATACAATTGCTATTGTTTATATAATGTAGTGACATTTTCTCTTAATCTTATACTAATTTCTATGACATTTATATAAGAAGAGACTTATGATCAACATAATTTTGCAAACTTTGAGAGAAATATGTCTTTCTACTGCGATAAAGTTATTATTTAGATTACATGTCACCAACATTTTCTCACCGTCGCGCTGAATCAGACAGGCTTTAGAGCGGTAGATAACTTCGACTTCACGCAGGCCGTTAGCTTCAAGCGTCGCGCCGGTAGAGACCAAATTTAAATGTTGAGGTAAATAGCTATTTTCTCTCTTCTTTTCCTTTAGTTGGAATTTGCACAAGAAAATGTTTTTCCACACACTTTAGCGTTTTTTCCTAAAGTGTTGCCAGCGATAAACGGCAGCCGCCGAAGTCAAGACGGCGCAGGGTTAAATAGCGTGGATCTTCGCCCTGTGCGCGGCGGTTGAGTAGCTCTTCTTCCAGCACGTTTTCGCCGATAATACCGAGATCGACCACGCCATCCATTACCAGACCCGGAATGTCATCATCACGCACGCGCAGGATATCAATCGGCATGTTTTCCGCCATCGCAATCAGGCGCTGAGTGTGTAAATTAATTTTTATGCCGCAGCGGGCCAGCAATTCTCGTGAATCATCGCTTAAACGGCCTGATTTCTGAATAGCTATGCGTAAGCGGGTGTTGTCTAATCTGCAGATTACTTTCCTAAATTTGCATATAGAATTGTCAAGCGCAAATCCGACGTCGATTCCGCGGCGGATGGGTCATTCTAGGTCATTCTACCAATTTTATTTGAGACCAGGTTTATTCAACCGGTAACATAGAAATATTCATACAATTAAGCTTCTATGGCCAAGTTGGTAAGGCGCCACACTAGTAATGTGGAGATCATCGGTTCAAATCCGATTGGAAGCATTTTTTATCACGTTATTCGGTGACACCCAGGTTGCCGCCGCGTTCGCGTCCATCGTCATCTGAAAAATAATGAATATTTTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTTTACTTTTGTATATTAAATTTACCGTTATTCGGAGATCTCTTACGGCTTATGATTTTCTTTACATTCCAGGCCGCCTTTTGATGACATTCAGTGCGTCACGACGCGAATAGGGCTTTGAAAGCCTTAATGTGGAGTAAATAAATATATGAATGCAATAAGTAAACATAAGACGAGTTCTTCTATATCCGGTGTACCTAATAATATAGATTTTGCCAACAATGGAATCTCATTACCCATCCCAAGATAAAATATTGATCAAGGCATGGGCCACCAACATAGGAATACGCAAAATAGGATTGATATTTTTACATATTATATTATCTTTCTTCTTCTGTCTCAAATATAGCTCTTCTCCTATTTAAAAAATTATCAATTCTTGAATCTCAACTTCCATTAAATTTGATGATTGTTTCCCAATATCGATGTCATCTTTTTACACTGTAAATTATAATTTACTAGGAACATGAATACTGATCAATGGATGATAGTTGATTCTTATTGCAATAAAAAGTACTTTCTAAGACGCTAGGACTTCTAAACACAGATATTATGAAAATGCCAAAAAATACAGAATATTTTCCTTCTTTTCACTATAACAAGTAGTAAGAAATTAAGCAATATGGTAATTTGAGCTTATCAGATAAAAGATCCTGGTAGTCTTGCAAAGGTGTGAGTACTTTTGATGAAACTAAGTTTCAGCGCAGCTATTTCAATTTTGGAACCTTTAAAGAAAGTTTAATCGAAACATTGAACGATCCAAAGCACAACTTTGTAGCGGGAAGTCACAAACAAATATAAAAAGTGTTAATAAACTGCCATCCTATGTTAGAATTGTGAAACTTTTTGGACTAATAAAAGGTTTTCAACATGGTGTTCCAAAGCACTTATTTCTAGGTTAGCTATGGTAACGAGTTTTAACGTAATCAAAACAATATACAATTCAAGATCTGTATCATTTACACTAACACACCCCATTTTAATAGAGCTGAGTACTTCCTACATTTCTTTTGCATATGATATCGATACTGCATCATTTCTGTAGTCGTAAAATTTCCTTCTCATTAAGACACCGAAAATAGCTAAGGATTATAACAAAACAAACAAATGCTCAAACAAAACTTTTAGTAAATCTTTTGAAATGTAAAAGTTACTTAGAATAACTAAAAAAAAAATATTACGCTATACCATATATTTCCATTCCCAATTTTATCGTTTTATAATAATATTCAGTTTCAAGAACAATGAATATATTTCAAACATTACATGCACTCTTGGGAGTAAGCTGCATCATAATAAATTAAAATTAACGCTTTAGAAAAAGAAATTTTGAGGTTAAAGGATGGAGG")

pRG727 <- c(pRG727_up, pRG727_down)

# find corresponding coordinates in genome
start <- start(matchPattern(pattern = pRG727[1:50], subject = genome[["chrIII"]]))
end <- end(matchPattern(pattern = pRG727[(length(pRG727)-50):length(pRG727)], subject = genome[["chrIII"]]))

# replace
genome[["chrIII"]][start:end] <- pRG727

# are +/- 1 kb around DSB site present?
HOcs_flanking <- DNAString(x = "GACATTTTCTCTTAATCTTATACTAATTTCTATGACATTTATATAAGAAGAGACTTATGATCAACATAATTTTGCAAACTTTGAGAGAAATATGTCTTTCTACTGCGATAAAGTTATTATTTAGATTACATGTCACCAACATTTTCTCACCGTCGCGCTGAATCAGACAGGCTTTAGAGCGGTAGATAACTTCGACTTCACGCAGGCCGTTAGCTTCAAGCGTCGCGCCGGTAGAGACCAAATTTAAATGTTGAGGTAAATAGCTATTTTCTCTCTTCTTTTCCTTTAGTTGGAATTTGCACAAGAAAATGTTTTTCCACACACTTTAGCGTTTTTTCCTAAAGTGTTGCCAGCGATAAACGGCAGCCGCCGAAGTCAAGACGGCGCAGGGTTAAATAGCGTGGATCTTCGCCCTGTGCGCGGCGGTTGAGTAGCTCTTCTTCCAGCACGTTTTCGCCGATAATACCGAGATCGACCACGCCATCCATTACCAGACCCGGAATGTCATCATCACGCACGCGCAGGATATCAATCGGCATGTTTTCCGCCATCGCAATCAGGCGCTGAGTGTGTAAATTAATTTTTATGCCGCAGCGGGCCAGCAATTCTCGTGAATCATCGCTTAAACGGCCTGATTTCTGAATAGCTATGCGTAAGCGGGTGTTGTCTAATCTGCAGATTACTTTCCTAAATTTGCATATAGAATTGTCAAGCGCAAATCCGACGTCGATTCCGCGGCGGATGGGTCATTCTAGGTCATTCTACCAATTTTATTTGAGACCAGGTTTATTCAACCGGTAACATAGAAATATTCATACAATTAAGCTTCTATGGCCAAGTTGGTAAGGCGCCACACTAGTAATGTGGAGATCATCGGTTCAAATCCGATTGGAAGCATTTTTTATCACGTTATTCGGTGACACCCAGGTTGCCGCCGCGTTCGCGTCCATCGTCATCTGAAAAATAATGAATATTTTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTTTACTTTTGTATATTAAATTTACCGTTATTCGGAGATCTCTTACGGCTTATGATTTTCTTTACATTCCAGGCCGCCTTTTGATGACATTCAGTGCGTCACGACGCGAATAGGGCTTTGAAAGCCTTAATGTGGAGTAAATAAATATATGAATGCAATAAGTAAACATAAGACGAGTTCTTCTATATCCGGTGTACCTAATAATATAGATTTTGCCAACAATGGAATCTCATTACCCATCCCAAGATAAAATATTGATCAAGGCATGGGCCACCAACATAGGAATACGCAAAATAGGATTGATATTTTTACATATTATATTATCTTTCTTCTTCTGTCTCAAATATAGCTCTTCTCCTATTTAAAAAATTATCAATTCTTGAATCTCAACTTCCATTAAATTTGATGATTGTTTCCCAATATCGATGTCATCTTTTTACACTGTAAATTATAATTTACTAGGAACATGAATACTGATCAATGGATGATAGTTGATTCTTATTGCAATAAAAAGTACTTTCTAAGACGCTAGGACTTCTAAACACAGATATTATGAAAATGCCAAAAAATACAGAATATTTTCCTTCTTTTCACTATAACAAGTAGTAAGAAATTAAGCAATATGGTAATTTGAGCTTATCAGATAAAAGATCCTGGTAGTCTTGCAAAGGTGTGAGTACTTTTGATGAAACTAAGTTTCAGCGCAGCTATTTCAATTTTGGAACCTTTAAAGAAAGTTTAATCGAAACATTGAACGATCCAAAGCACAACTTTGTAGCGGGAAGTCACAAACAAATATAAAAAGTGTTAATAAACTGCCATCCTATGTTAGAATTGTGAAACTTTTTGGACTAATAAAAGGTTTTCAACATGGTGTTCCAAAGCACTTATTTCTAGGTTAGCTATGGTAACGAGTTTTAACGTAATCAAAACAATATACAATTCAAGATCTGTATCATTTACACTAACACACCCCAT")
matchPattern(pattern = HOcs_flanking, subject = genome[["chrIII"]])

# are there two HOcs now on chrIII?
matchPattern(pattern = "ttCTTTTaGTTTCAGCTTTCCGCAACAGTATAATTTTATAAAcCCTGGtt", subject = genome[["chrIII"]])

# export modified reference genome to fast file
export(object = genome, con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Data/S_cerevisiae_reference_genomes/S288C_R64-2-1_W303_SNPs_MATa_hml_hmr_LSY4754-3A.fasta")
