# info --------------------------------------------------------------------
# purpose: modify S288C reference genome to improve mapping of W303-derived reads; add strain-specific mutations
# author: Robert Gnuegge (robert.gnuegge@gmail.com)
# date: 01/28/22
# version: 1.0

# preamble ----------------------------------------------------------------

# set wd to this file's location
wd_path <- dirname(rstudioapi::getActiveDocumentContext()$path)
setwd(wd_path)

# libraries
library(GenomicRanges)
library(Biostrings)
library(rtracklayer)

source(file = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Analysis_scripts/Misc/S_cerevisiae_SrfI_cut_sites.R")


# load S288C reference genome ---------------------------------------------
# sequence downloaded from http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/ on 05/11/20

genome <- import(con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/S288C_reference_sequence_R64-2-1_20150113.fasta")

# change names
names(genome)
names(genome) <- c(paste0("chr", as.roman(1:16)), "chrM")
names(genome)


# add 2 micron plasmid ----------------------------------------------------
# sequence downloaded from https://www.ncbi.nlm.nih.gov/nuccore/CM001822.1 on 05/11/20

micron <- import(con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/2micron_W303.fasta")

# the 2 micron plasmid is circular
# to support read mapping near its 3' end, add a sequence stretch from its 5' end
read_length <- 75
micron$micron <- c(micron$micron, micron$micron[1:(read_length - 1)])

# add to genome
genome <- DNAStringSet(x = c(as.character(genome), as.character(micron)))
seqinfo(genome)

# inject W303 SNPs --------------------------------------------------------
# SNPs downloaded from https://www.g3journal.org/highwire/filestream/485032/field_highwire_adjunct_files/0/FileS1.xlsx on 05/11/20
# (Matheson et al., 2017)

SNPs <- read.csv(file = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Resources/S_cerevisiae_genome/W303_SNPs.csv", header = TRUE, stringsAsFactors = FALSE)

# change chrmt to chrM
SNPs$Chromosome[SNPs$Chromosome == "chrmt"] <- "chrM"

# check that the S288C nts in the SNPs file are consistent with the genome sequence
all(sapply(X = names(genome)[1:17], FUN = function(c) {as.character(genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]]) == paste0(SNPs$Base.in.S288c[SNPs$Chromosome == c], collapse = "")}))

# substitute with W303 SNPs
for(c in c(paste0("chr", as.roman(1:16)), "chrM")){
  genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]] <- DNAString(paste0(SNPs$Base.in.W303[SNPs$Chromosome == c], collapse = ""))
}

# check that SNPs have been injected
all(sapply(X = names(genome)[1:17], FUN = function(c) {as.character(genome[[c]][SNPs$Position.in.S288c[SNPs$Chromosome == c]]) == paste0(SNPs$Base.in.W303[SNPs$Chromosome == c], collapse = "")}))


# change MATalpha to MATa -------------------------------------------------
# the strains used for S1-seq library preparation are all MATa

# correct HMRa1 stk mutation (see https://www.yeastgenome.org/locus/S000029699#paragraph for details)
matchPattern(pattern = DNAString("CGCAACAGTAAAATTTTATAA"), subject = genome[["chrIII"]])
matchPattern(pattern = DNAString("CGCAACAGTATAATTTTATAA"), subject = genome[["chrIII"]])
# already corrected during the W303 SNP injection above

# define MATalpha Z region (coordinates from https://www.yeastgenome.org/locus/S000029699)
Z1 <- genome[["chrIII"]][200851:201089]

# find Z regions in HML, MAT, and HMR
all.Z1 <- matchPattern(pattern = Z1, subject = genome[["chrIII"]], max.mismatch = 1)

# turns out that there is a 1 bp difference between HML/MAT and HMR (C->T)
mismatch(pattern = Z1, x = all.Z1)

# define MATalpha X region
X <- genome[["chrIII"]][199401:200103] 

# find X regions in HML, MAT, and HMR
all.X <- matchPattern(pattern = X, subject = genome[["chrIII"]], max.mismatch = 2, with.indels = TRUE)

# MATalpha has a 2 bp deletion at position 199512:199513 (in a poly-TA stretch)
mismatch(pattern = X, x = all.X)  # gives mismatches for HML and HMR at relative position > 111

# let's only use the first 111 bp of the MATalpha X
X <- genome[["chrIII"]][199401:199511] 

# find X regions in HML, MAT, and HMR
all.X <- matchPattern(pattern = X, subject = genome[["chrIII"]], max.mismatch = 0)

# MATalpha has a 2 bp deletion at position 199512:199513 (in a poly-TA stretch)
mismatch(pattern = X, x = all.X)  # gives mismatches for HML and HMR at relative position > 111

# check size difference of regions
end(all.Z1[2]) - start(all.X[2]) - ( end(all.Z1[3]) - start(all.X[3]) )
# MATalpha X-Z region is 103 bp longer than MATa X-Z region

# replace MATalpha with HMR and add 103 Ns downstream of TAF1 gene (ca. 5 kb downstream of HO cut site)
TAF1.end <- 206260  # upstream of YCR043C, which is downstream of TAF1 (https://www.yeastgenome.org/locus/YCR043C)
MATalpha <- genome[["chrIII"]][start(all.X[2]):TAF1.end]

HMR <- genome[["chrIII"]][start(all.X[3]):end(all.Z1[3])]
TAF1 <- genome[["chrIII"]][(end(all.Z1[2]) + 1):TAF1.end]
Ns <- DNAString(x = paste0(rep("N", 103), collapse = ""))
replacement <- c(HMR, TAF1, Ns)

# sanity check
length(MATalpha) == length(replacement)

# replace
genome[["chrIII"]][start(all.X[2]):TAF1.end] <- replacement

# check if HMR sequence was copied to MAT
matchPattern(pattern = HMR, subject = genome[["chrIII"]])

HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = genome[["chrIII"]])

# HML deletion ------------------------------------------------------------
# the HML deletion was generated by C. Zierhut (https://www.ncbi.nlm.nih.gov/pubmed/18511906 and personal communication)
# HML was replaced by URA3, which was then in turn replaced by a PCR fragment derived from pRS306 (ori) using the following primers
OCZ330 <- DNAString(x = "gacaagtagcgcagttattttcttattttcatCATCAGAGCAGATTGTACTGAGAGTGCAtaattgcgttgcgctcactg")
OCZ331 <- reverseComplement(x = DNAString(x = "CAAATTTACTAGCTcgagagttagttaatccttactaagtgaagaaaagcaactatagtcTCAAGAACTCTGTAGCACCG"))

# find replaced genome region
HML.start <- start(matchPattern(pattern = OCZ330[1:32], subject = genome[["chrIII"]]))
HML.end <- end(matchPattern(pattern = OCZ331[21:80], subject = genome[["chrIII"]]))
HML <- genome[["chrIII"]][HML.start:HML.end]

# find amplified pRS306 region
# pRS306 sequence downloaded from https://www.ncbi.nlm.nih.gov/nuccore/U03438.1/ on 05/11/20
pRS306 <- import(con = "~/Research/Resources/pRS306.fasta")[[1]]
ori.start <- start(matchPattern(pattern = OCZ330[61:80], subject = pRS306))
ori.end <- end(matchPattern(pattern = OCZ331[1:20], subject = pRS306))
ori <- pRS306[ori.start:ori.end]
# expand with primer overhangs
replacement <- c(OCZ330[1:60], ori, OCZ331[21:80])

# length of replacement is 3495 bp shorter than replaced region
length(HML) - length(replacement)
# let's fill up with Ns in the center of the replacement
replacement <- c(replacement[1:(round(0.5 * length(replacement)) - 1)],
                 DNAString(x = paste0(rep("N", 3495), collapse = "")),
                 replacement[round(0.5 * length(replacement)):length(replacement)])

# sanity check
length(HML) == length(replacement)

# replace
genome[["chrIII"]][HML.start:HML.end] <- replacement

# check if ori parts were inserted into genome
matchPattern(pattern = ori[1:floor(0.5 * length(ori) - 1)], subject = genome[["chrIII"]])
matchPattern(pattern = ori[ceiling(0.5 * length(ori) + 1):length(ori)], subject = genome[["chrIII"]])
matchPattern(pattern = HML, subject = genome[["chrIII"]])


# HMR deletion ------------------------------------------------------------
# the HMR deletion was generated by C. Zierhut (https://www.ncbi.nlm.nih.gov/pubmed/18511906 and personal communication).
# HMR was replaced by URA3, which was then in turn replaced by a PCR fragment derived from pRS306 (bla) using the following primers
OCZ332 <- DNAString(x = "gatttaagcgtgcgtGAAGATAACACTACAATCCATTTTAAAGCAACATCCACATTGAGTcgttcatccatagttgcctg")
OCZ333 <- reverseComplement(x = DNAString(x = "ggtatatagaatataCTAGAAGTTCTCCTCGAGGATATAGGAATCCTCAAAAGGGAATCTttcaacatttccgtgtcgcc"))

# find replaced genome region
HMR.start <- start(matchPattern(pattern = OCZ332[1:60], subject = genome[["chrIII"]]))
HMR.end <- end(matchPattern(pattern = OCZ333[21:80], subject = genome[["chrIII"]]))
HMR <- genome[["chrIII"]][HMR.start:HMR.end]

# find amplified pRS306 region
pRS306 <- import(con = "~/Research/Resources/pRS306.fasta")[[1]]
bla.start <- start(matchPattern(pattern = OCZ332[61:80], subject = pRS306))
bla.end <- end(matchPattern(pattern = OCZ333[1:20], subject = pRS306))
bla <- pRS306[bla.start:bla.end]

# expand with primer overhangs
replacement <- c(OCZ332[1:60], bla, OCZ333[21:80])

# length of replacement is 3341 bp shorter than replaced region
length(HMR) - length(replacement)
# let's fill up with Ns in the center of the replacement
replacement <- c(replacement[1:(round(0.5 * length(replacement)) - 1)],
                 DNAString(x = paste0(rep("N", 3341), collapse = "")),
                 replacement[round(0.5 * length(replacement)):length(replacement)])

# sanity check
length(HMR) == length(replacement)

# replace
genome[["chrIII"]][HMR.start:HMR.end] <- replacement

# check if bla parts were inserted into the genome
matchPattern(pattern = bla[1:floor(0.5 * length(bla) - 1)], subject = genome[["chrIII"]])
matchPattern(pattern = bla[ceiling(0.5 * length(bla) + 1):length(bla)], subject = genome[["chrIII"]])
matchPattern(pattern = HMR, subject = genome[["chrIII"]])

# LSY4810-41D modifications ===============================================
LSY4810_41D <- genome

# -----------------------------------------------------------------------------
pRG721 <- DNAString(x = "CCTCCTCTAGTACACTCTATATTTTTTTATGCCTCGGTAATGATTTTCATTTTTTTTTTTCCACCTAGCGGATGACTCTTTTTTTTTCTTAGCGATTGGCATTATCACATAATGAATTATACATTATATAAAGTAATGTGATTTCTTCGAAGAATATACTAAAAAATGAGCAGGCAAGATAAACGAAGGCACATATGAGGAGGGCTTTTGTAGAAAGAAATACGAACGAAACGAAAATCAGCGTTGCCATCGCTTTGGACAAAGCTCCCTTACCTGAAGAGTCGAATTTTATTGATGAACTTATAACTTCCAAGCATGCAAACCAAAAGGGAGAACAAGTAATCCAAGTAGACACGGGAATTGGATTCTTGGATCACATGTATCATGCACTGGCTAAACATGCAGGCTGGAGCTTACGACTTTACTCAAGAGGTGATTTAATCATCGATGATCATCACACTGCAGAAGATACTGCTATTGCACTTGGTATTGCATTCAAGCAGGCTATGGGTAACTTTGCCGGTGTTAAAAGATTTGGACATGCTTATTGTCCACTTGACGAAGCTCTTTCCAGAAGCGTAGTTGACTTGTCGGGACGGCCCTATGCTGTTATCGATTTGGGATTAAAGCGTGAAAAGGTTGGGGAATTGTCCTGTGAAATGATCCCTCACTTACTATATTCCTTTTCGGTAGCTGCTGGAATTACTTTGCATGTTACCTGCTTATATGGTAGTAATGACCATCATCGTGCTGAAAGCGCTTTTAAATCTCTGGCTGTTGCCATGCGCGCGGCTACTTCTCTTACTGGAAGTTCTGAAGTCCCAAGCACGAAGGGAGTGTTGTAAAGATGAATTGGATTATGTCAGGAAAAGAACGACAATTTTGCATCCAAATTGTCTAAATTTTAGAGTTGCTTGAAAACAATAGAACCTTACTTGCTTTATAATTACGTTAATTAGAAGCGTTATCTCGTGAAGGAATATAGTACGTAGCCGTATAAATTGAATTGAATGTTCAGCTTATAGAATAGAGACACTTTGCTGTTCAATGCGTCGTCACTTACCATACTCACTTTATTATACGACTTTAAGTATAAACTCCGCGGTTATGGTAAAATTAATGATGCACAAACGTCCGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTGGAGCTCCACCGCGGTGGCTCTAGAATGAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAAGAATTCGATATCAAGCTTttttCTTTTaGTTTCAGCTTTCCGCaacaGTATAATTTTATAAAcCCTGGtTACCCATACGATGTTCCAGATTACGCTtaaCTCGAGACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTTGgtaccCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGACACCGATTATTTAAAGCTGCAGCATACGATATATATACATGTGTATATATGTATACCTATGAATGTCAGTAAGTATGTATACGAACAGTATGATACTGAAGATGACAAGGTAATGCATCATTCTATACGTGTCATTCTGAACGAGG")

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG721[1:50], subject = LSY4810_41D[["chrXV"]]))
end <- end(matchPattern(pattern = pRG721[(length(pRG721)-50):length(pRG721)], subject = LSY4810_41D[["chrXV"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG721) - (end - start)
LSY4810_41D[["chrXV"]][(start - nt_added + 1):end] <- pRG721

# is the sequence present now?
matchPattern(pattern = pRG721, subject = LSY4810_41D[["chrXV"]])

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4810_41D[["chrXV"]])

# mask endogenous sequences corresponding to reporter parts
T_CYC1 <- DNAString(x = "ACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTT")
matchPattern(pattern = T_CYC1, subject = LSY4810_41D[["chrX"]], max.mismatch = 2)
LSY4810_41D[["chrX"]][matchPattern(pattern = T_CYC1, subject = LSY4810_41D[["chrX"]], max.mismatch = 2)] <- DNAString(x = paste0(rep("N", length(T_CYC1)), collapse = ""))
matchPattern(pattern = T_CYC1, subject = LSY4810_41D[["chrX"]], max.mismatch = 2)

# -----------------------------------------------------------------------------
pRG723_up <- DNAString(x = "CCCTTAGCTCTCATTATTTTTTGCTTTTTCTCTTGAGGTCACATGATCGCAAAATGGCAAATGGCACGTGAAGCTGTCGATATTGGGGAACTGTGGTGGTTGGCAAATGACTAATTAAGTTAGTCAAGGCGCCATCCTCATGAAAACTGTGTAACATAATAACCGAAGTGTCGAAAAGGTGGCACCTTGTCCAATTGAACACGCTCGATGAAAAAAATAAGATATATATAAGGTTAAGTAAAGCGTCTGTTAGAAAGGAAGTTTTTCCTTTTTCTTGCTCTCTTGTCTTTTCATCTACTATTTCCTTCGTGTAATACAGGGTCGTCAGATACATAGATACAATTCTATTACCCCCATCCATACAATGCCATCTCATTTCGATACTGTTCAACTACACGCCGGTCAAGAGAACCCTGGTGACAATGCTCACAGATCCAGAGCTGTACCAATTTACGCCACCACTTCTTATGTTTTCGAAAACTCTAAGCATGGTTCGCAATTGTTTGGTTTAGAAGTTCCAGGTTACGTCTATTCCCGTTTCCAAAACCCAACCAGTAATGTTTTGGAAGAAAGAATTGCTGCTTTAGAAGGTGGTGCTGCTGCTTTGGCTGTTTCCTCCGGTCAAGCCGCTCAAACCCTTGCCATCCAAGGTTTGGCACACACTGGTGACAACATCGTTTCCACTTCTTACTTATACGGTGGTACTTATAACCAGTTCAAAATCTCGTTCAAAAGATTTGGTATCGAGGCTAGATTTGTTGAAGGTGACAATCCAGAAGAATTTGAAAAGGTCTTTGATGAAAGAACCAAGGCTGTTTATTTGGAAACCATTGGTAATCCAAAGTACAATGTTCCGGATTTTGAAAAAATTGTTGCAATTGCTCACAAACACGGTATTCCAGTTGTCGTTGACAACACATTTGGTGCCGGTGGTTACTTCTGTCAGCCAATTAAATACGGTGCTGATATTGTAACACATTCTGCTACCAAATGGATTGGTGGTCATGGTACTACTATCGGTGGTATTATTGTTGACTCTGGTAAGTTCCCATGGAAGGACTACCCAGAAAAGTTCCCTCAATTCTCTCAACCTGCCGAAGGTTATCACGGTACTATCTACAATGAAGCCTACGGTAACTTGGCATACATCGTTCATGTTAGAACTGAACTATTAAGAGATTTGGGTCCATTGATGAACCCATTTGCCTCTTTCTTGCTACTACAAGGTGTTGAAACATTATCTTTGAGAGCTGAAAGACACGGTGAAAATGCATTGAAGTTAGCCAAATGGTTAGAACAATCCCCATACGTATCTTGGGTTTCATACCCTGGTTTAGCATCTCATTCTCATCATGAAAATGCTAAGAAGTATCTATCTAACGGTTTCGGTGGTGTCTTATCTTTCGGTGTAAAAGACTTACCAAATGCCGACAAGGAAACTGACCCATTCAAACTTTCTGGTGCTCAAGTTGTTGACAATTTAAAGCTAGCCTCTAACTTGGCCAATGTTGGTGATGCCAAGACCTTAGTCATTGCTCCATACTTCACTACCCACAAACAATTAAATGACAAAGAAAAGTTGGCATCTGGTGTTACCAAGGACTTAATTCGTGTCTCTGTTGGTATCGAATTTATTGATGACATTATTGCAGACTTCCAGCAATCTTTTGAAACTGTTTTCGCTGGCCAAAAACCATGAGTGTGCGTAATGAGTTGTAAAATTATGTATAAACCTACTTTCTCTCACAAGTACTATACTTTTATAAAACGAACTTTATTGAAATGAATATCCTTTTTTTCCCTTGTTACATGTCGTGACTCGTACTTTGAACCTAAATTGTTCTAACATCAAAGAACAGTGTTAATTCGCAGTCGAGAAGAAAAATATGGTGAACAAGACTCATCTACTTCATGAGACTACTTTACGCCTCCTATAAAGCTGTCACACTGGATAAATTTATTGTAGGACCAAGTTACAAAAGAGGATGATGGAGGTTTCTTTACAATAAAGAAGCACATGTGTGTTAACGTTTTTAGTATTTGCTTGTTATGTAAATCAGGAAAACTTCGCGGGATTTGGTTGGATGCTACTTTCCATACAATAAATATTATAGATCAAAAAAGCCAAATTACAAGTAAAGATTAGTAAAGCTGTTGCAATTCCATCGTTGATAAAAATGTTAGTTATTAAATATAAAAGTCAGAATAGGTGAACTTGGATTTAATTGTTGGCATTTCGTTGCTGCTAGAGGCCATAATATTAGATAGCCAGGACATACTAGGTCTCCTCGTGGTATAGGAATCCATAAAATGGAATTGGTGACGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTG")
pRG723_down <- DNAString(x = "GAGCTCCCGCTCTAGACAGTTCGAGTTTATCATTATCAATACTGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAAGAACTTAGTTTCGAATAAACACACATAAACAAACAAAACTAGAATGGTTTCTGAATTGATTAAAGAAAACATGCATATGAAGTTGTATATGGAAGGTACTGTTAACAATCATCATTTCAAGTGTACATCTGAAGGTGAAGGTAAACCATATGAAGGTACTCAAACTATGAGAATTAAAGCTGTTGAAGGTGGTCCATTGCCATTTGCTTTTGATATTTTGGCTACTTCTTTCATGTATGGTTCTAAGACTTTCATTAACCATACTCAAGGTATTCCAGATTTTTTCAAGCAATCTTTTCCAGAAGGTTTTACTTGGGAAAGAGTTACTACTTACGAAGATGGTGGTGTTTTGACTGCTACTCAAGATACTTCTTTGCAAGATGGTTGTTTGATTTACAACGTTAAGATTAGAGGTGTTAATTTTCCATCTAATGGTCCAGTTATGCAAAAAAAGACTTTGGGTTGGGAAGCATCTACTGAAACTTTGTATCCAGCTGATGGTGGTTTGGAAGGTAGAGCTGATATGGCTTTGAAATTGGTTGGTGGTGGTCATTTGATTTGTAACTTGAAAACTACTTACAGATCAAAAAAGCCAGCTAAGAATTTGAAAATGCCAGGTGTTTATTACGTTGATAGAAGATTGGAAAGAATTAAAGAAGCTGATAAAGAAACTTACGTTGAACAACATGAAGTTGCTGTTGCAAGATATTGTGATTTGCCATCTAAATTGGGTCATAGATAAgaattcTTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTGaagcttATCGATACCGTCGACCTCGAGttCTTTTaGTTTCAGCTTTCCGCaacaGTATAATTTTATAAAcCCTGGttGGTACCCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGTCAGATATAGTCGGATTGCCCTTTTAAGCAATCAATAGTGTTTTATTTGCAACAATGTCGTCATAGTTTAATATGTCCTATAAGATGTTAACTTGCTCAACATTCAACAAAGTTTGGTCTCTTGGCCCAGTTGGTTAAGGCACCGTGCTAATAACGCGGGGATCAGCGGTTCGATCCCGCTAGAGACCATTTATTTTTAAATTCGACCGTCTGACAGACGGTGCACTTATTACTTAATTTTTTTTTTTTTTTAATCCTCGTTAAACTGAGTACATAATCAGCAACAATATAATATAAAAGTTTTGCTTTTTTATGTACGTACAATAGGGGTACTTAAATAAATAAATAAACAATAAAACAATCAATAAATAAATAATATACGATCATAAATTACGAGATATATTTCATAGTCTTGACTTCTCCAGCCGAACGTCATTAAATTTAAGGCAGAAAAGATAAACAAGGG")
pRG723 <- c(pRG723_up, pRG723_down)

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG723[1:50], subject = LSY4810_41D[["chrXII"]]))
end <- end(matchPattern(pattern = pRG723[(length(pRG723)-50):length(pRG723)], subject = LSY4810_41D[["chrXII"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG723) - (end - start)
LSY4810_41D[["chrXII"]][(start - nt_added + 1):end] <- pRG723

# is the sequence present now?
matchPattern(pattern = pRG723, subject = LSY4810_41D[["chrXII"]])

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4810_41D[["chrXII"]])

# mask endogenous sequences corresponding to reporter parts
P_TDH3 <- DNAString(x = "CAGTTCGAGTTTATCATTATCAATACTGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAAGAACTTAGTTTCGAATAAACACACATAAACAAACAAA")
matchPattern(pattern = reverseComplement(P_TDH3), subject = LSY4810_41D[["chrVII"]])
LSY4810_41D[["chrVII"]][matchPattern(pattern = reverseComplement(P_TDH3), subject = LSY4810_41D[["chrVII"]])] <- DNAString(x = paste0(rep("N", length(P_TDH3)), collapse = ""))
matchPattern(pattern = reverseComplement(P_TDH3), subject = LSY4810_41D[["chrVII"]])

T_CYC2 <- DNAString(x = "TTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTG")
matchPattern(pattern = T_CYC2, subject = LSY4810_41D[["chrXV"]])
LSY4810_41D[["chrXV"]][matchPattern(pattern = T_CYC2, subject = LSY4810_41D[["chrXV"]])] <- DNAString(x = paste0(rep("N", length(T_CYC2)), collapse = ""))
matchPattern(pattern = T_CYC2, subject = LSY4810_41D[["chrXV"]])

# find mKate2 sequence
mKate2 <- DNAString(x = "ATGGTTTCTGAATTGATTAAAGAAAACATGCATATGAAGTTGTATATGGAAGGTACTGTTAACAATCATCATTTCAAGTGTACATCTGAAGGTGAAGGTAAACCATATGAAGGTACTCAAACTATGAGAATTAAAGCTGTTGAAGGTGGTCCATTGCCATTTGCTTTTGATATTTTGGCTACTTCTTTCATGTATGGTTCTAAGACTTTCATTAACCATACTCAAGGTATTCCAGATTTTTTCAAGCAATCTTTTCCAGAAGGTTTTACTTGGGAAAGAGTTACTACTTACGAAGATGGTGGTGTTTTGACTGCTACTCAAGATACTTCTTTGCAAGATGGTTGTTTGATTTACAACGTTAAGATTAGAGGTGTTAATTTTCCATCTAATGGTCCAGTTATGCAAAAAAAGACTTTGGGTTGGGAAGCATCTACTGAAACTTTGTATCCAGCTGATGGTGGTTTGGAAGGTAGAGCTGATATGGCTTTGAAATTGGTTGGTGGTGGTCATTTGATTTGTAACTTGAAAACTACTTACAGATCAAAAAAGCCAGCTAAGAATTTGAAAATGCCAGGTGTTTATTACGTTGATAGAAGATTGGAAAGAATTAAAGAAGCTGATAAAGAAACTTACGTTGAACAACATGAAGTTGCTGTTGCAAGATATTGTGATTTGCCATCTAAATTGGGTCATAGATAA")
matchPattern(pattern = mKate2, subject = LSY4810_41D[["chrXII"]])

# -----------------------------------------------------------------------------
# export to fasta file
export(object = LSY4810_41D, con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Data/S_cerevisiae_reference_genomes/S288C_R64-2-1_W303_SNPs_MATa_hml_hmr_LSY4810-41D.fasta")


# =============================================================================
# LSY4811-11B modifications
LSY4811_11B <- genome

# -----------------------------------------------------------------------------
pRG746 <- DNAString(x = "CCTCCTCTAGTACACTCTATATTTTTTTATGCCTCGGTAATGATTTTCATTTTTTTTTTTCCACCTAGCGGATGACTCTTTTTTTTTCTTAGCGATTGGCATTATCACATAATGAATTATACATTATATAAAGTAATGTGATTTCTTCGAAGAATATACTAAAAAATGAGCAGGCAAGATAAACGAAGGCACATATGAGGAGGGCTTTTGTAGAAAGAAATACGAACGAAACGAAAATCAGCGTTGCCATCGCTTTGGACAAAGCTCCCTTACCTGAAGAGTCGAATTTTATTGATGAACTTATAACTTCCAAGCATGCAAACCAAAAGGGAGAACAAGTAATCCAAGTAGACACGGGAATTGGATTCTTGGATCACATGTATCATGCACTGGCTAAACATGCAGGCTGGAGCTTACGACTTTACTCAAGAGGTGATTTAATCATCGATGATCATCACACTGCAGAAGATACTGCTATTGCACTTGGTATTGCATTCAAGCAGGCTATGGGTAACTTTGCCGGTGTTAAAAGATTTGGACATGCTTATTGTCCACTTGACGAAGCTCTTTCCAGAAGCGTAGTTGACTTGTCGGGACGGCCCTATGCTGTTATCGATTTGGGATTAAAGCGTGAAAAGGTTGGGGAATTGTCCTGTGAAATGATCCCTCACTTACTATATTCCTTTTCGGTAGCTGCTGGAATTACTTTGCATGTTACCTGCTTATATGGTAGTAATGACCATCATCGTGCTGAAAGCGCTTTTAAATCTCTGGCTGTTGCCATGCGCGCGGCTACTTCTCTTACTGGAAGTTCTGAAGTCCCAAGCACGAAGGGAGTGTTGTAAAGATGAATTGGATTATGTCAGGAAAAGAACGACAATTTTGCATCCAAATTGTCTAAATTTTAGAGTTGCTTGAAAACAATAGAACCTTACTTGCTTTATAATTACGTTAATTAGAAGCGTTATCTCGTGAAGGAATATAGTACGTAGCCGTATAAATTGAATTGAATGTTCAGCTTATAGAATAGAGACACTTTGCTGTTCAATGCGTCGTCACTTACCATACTCACTTTATTATACGACTTTAAGTATAAACTCCGCGGTTATGGTAAAATTAATGATGCACAAACGTCCGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTGGAGCTCCCGCTCTAGACAGTTCGAGTTTATCATTATCAATACTGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAAGAACTTAGTTTCGAATAAACACACATAAACAAACAAAACTAGAATGAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAAGaattcTTAGACAACACCCGCTTACGCATAGCTATTCAGAAATCAGGCCGTTTAAGCGATGATTCACGAGAATTGCTGGCCCGCTGCGGCATAAAAATTAATTTACACACTCAGCGCCTGATTaAGCTTttttCTTTTaGTTTCAGCTTTCCGCAACAGTATAATTTTATAAAcCCTGGtTACCCATACGATGTTCCAGATTACGCTtaaCTCGAGACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTTGgtaccCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGACACCGATTATTTAAAGCTGCAGCATACGATATATATACATGTGTATATATGTATACCTATGAATGTCAGTAAGTATGTATACGAACAGTATGATACTGAAGATGACAAGGTAATGCATCATTCTATACGTGTCATTCTGAACGAGG")

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG746[1:50], subject = LSY4811_11B[["chrXV"]]))
end <- end(matchPattern(pattern = pRG746[(length(pRG746)-50):length(pRG746)], subject = LSY4811_11B[["chrXV"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG746) - (end - start)
LSY4811_11B[["chrXV"]][(start - nt_added + 1):end] <- pRG746

# is the sequence present now?
matchPattern(pattern = pRG746, subject = LSY4811_11B[["chrXV"]])

# find Citrine-HOcs-HA sequence
Citrine <- DNAString(x = "ATGAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAAGaattcTTAGACAACACCCGCTTACGCATAGCTATTCAGAAATCAGGCCGTTTAAGCGATGATTCACGAGAATTGCTGGCCCGCTGCGGCATAAAAATTAATTTACACACTCAGCGCCTGATTaAGCTTttttCTTTTaGTTTCAGCTTTCCGCAACAGTATAATTTTATAAAcCCTGGtTACCCATACGATGTTCCAGATTACGCTtaa")
matchPattern(pattern = Citrine, subject = LSY4811_11B[["chrXV"]])

# mask endogenous sequences corresponding to reporter parts
T_CYC1 <- DNAString(x = "ACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTT")
matchPattern(pattern = T_CYC1, subject = LSY4811_11B[["chrX"]], max.mismatch = 2)
LSY4811_11B[["chrX"]][matchPattern(pattern = T_CYC1, subject = LSY4811_11B[["chrX"]], max.mismatch = 2)] <- DNAString(x = paste0(rep("N", length(T_CYC1)), collapse = ""))
matchPattern(pattern = T_CYC1, subject = LSY4811_11B[["chrX"]], max.mismatch = 2)

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4811_11B[["chrXV"]])

# -----------------------------------------------------------------------------
pRG722_up <- DNAString(x = "CCCTTAGCTCTCATTATTTTTTGCTTTTTCTCTTGAGGTCACATGATCGCAAAATGGCAAATGGCACGTGAAGCTGTCGATATTGGGGAACTGTGGTGGTTGGCAAATGACTAATTAAGTTAGTCAAGGCGCCATCCTCATGAAAACTGTGTAACATAATAACCGAAGTGTCGAAAAGGTGGCACCTTGTCCAATTGAACACGCTCGATGAAAAAAATAAGATATATATAAGGTTAAGTAAAGCGTCTGTTAGAAAGGAAGTTTTTCCTTTTTCTTGCTCTCTTGTCTTTTCATCTACTATTTCCTTCGTGTAATACAGGGTCGTCAGATACATAGATACAATTCTATTACCCCCATCCATACAATGCCATCTCATTTCGATACTGTTCAACTACACGCCGGTCAAGAGAACCCTGGTGACAATGCTCACAGATCCAGAGCTGTACCAATTTACGCCACCACTTCTTATGTTTTCGAAAACTCTAAGCATGGTTCGCAATTGTTTGGTTTAGAAGTTCCAGGTTACGTCTATTCCCGTTTCCAAAACCCAACCAGTAATGTTTTGGAAGAAAGAATTGCTGCTTTAGAAGGTGGTGCTGCTGCTTTGGCTGTTTCCTCCGGTCAAGCCGCTCAAACCCTTGCCATCCAAGGTTTGGCACACACTGGTGACAACATCGTTTCCACTTCTTACTTATACGGTGGTACTTATAACCAGTTCAAAATCTCGTTCAAAAGATTTGGTATCGAGGCTAGATTTGTTGAAGGTGACAATCCAGAAGAATTTGAAAAGGTCTTTGATGAAAGAACCAAGGCTGTTTATTTGGAAACCATTGGTAATCCAAAGTACAATGTTCCGGATTTTGAAAAAATTGTTGCAATTGCTCACAAACACGGTATTCCAGTTGTCGTTGACAACACATTTGGTGCCGGTGGTTACTTCTGTCAGCCAATTAAATACGGTGCTGATATTGTAACACATTCTGCTACCAAATGGATTGGTGGTCATGGTACTACTATCGGTGGTATTATTGTTGACTCTGGTAAGTTCCCATGGAAGGACTACCCAGAAAAGTTCCCTCAATTCTCTCAACCTGCCGAAGGTTATCACGGTACTATCTACAATGAAGCCTACGGTAACTTGGCATACATCGTTCATGTTAGAACTGAACTATTAAGAGATTTGGGTCCATTGATGAACCCATTTGCCTCTTTCTTGCTACTACAAGGTGTTGAAACATTATCTTTGAGAGCTGAAAGACACGGTGAAAATGCATTGAAGTTAGCCAAATGGTTAGAACAATCCCCATACGTATCTTGGGTTTCATACCCTGGTTTAGCATCTCATTCTCATCATGAAAATGCTAAGAAGTATCTATCTAACGGTTTCGGTGGTGTCTTATCTTTCGGTGTAAAAGACTTACCAAATGCCGACAAGGAAACTGACCCATTCAAACTTTCTGGTGCTCAAGTTGTTGACAATTTAAAGCTAGCCTCTAACTTGGCCAATGTTGGTGATGCCAAGACCTTAGTCATTGCTCCATACTTCACTACCCACAAACAATTAAATGACAAAGAAAAGTTGGCATCTGGTGTTACCAAGGACTTAATTCGTGTCTCTGTTGGTATCGAATTTATTGATGACATTATTGCAGACTTCCAGCAATCTTTTGAAACTGTTTTCGCTGGCCAAAAACCATGAGTGTGCGTAATGAGTTGTAAAATTATGTATAAACCTACTTTCTCTCACAAGTACTATACTTTTATAAAACGAACTTTATTGAAATGAATATCCTTTTTTTCCCTTGTTACATGTCGTGACTCGTACTTTGAACCTAAATTGTTCTAACATCAAAGAACAGTGTTAATTCGCAGTCGAGAAGAAAAATATGGTGAACAAGACTCATCTACTTCATGAGACTACTTTACGCCTCCTATAAAGCTGTCACACTGGATAAATTTATTGTAGGACCAAGTTACAAAAGAGGATGATGGAGGTTTCTTTACAATAAAGAAGCACATGTGTGTTAACGTTTTTAGTATTTGCTTGTTATGTAAATCAGGAAAACTTCGCGGGATTTGGTTGGATGCTACTTTCCATACAATAAATATTATAGATCAAAAAAGCCAAATTACAAGTAAAGATTAGTAAAGCTGTTGCAATTCCATCGTTGATAAAAATGTTAGTTATTAAATATAAAAGTCAGAATAGGTGAACTTGGATTTAATTGTTGGCATTTCGTTGCTGCTAGAGGCCATAATATTAGATAGCCAGGACATACTAGGTCTCCTCGTGGTATAGGAATCCATAAAATGGAATTGGTGACGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTG")
pRG722_down <- DNAString(x = "GAGCTCCACCGCGGTGGCGGCCGCTCTAGAGAAGCGGGTAAGCTGCCACAGCAATTAATGCACAACATTTAACCTACATTCTTCCTTATCGGATCCTCAAAACCCTTAAAAACATATGCCTCACCCTAACATATTTTCCAATTAACCCTCAATATTTCTCTGTCACCCGGCCTCTATTTTCCATTTTCTTCTTTACCCGCCACGCGTTTTTTTCTTTCAAATTTTTTTCTTCCTTCTTCTTTTTCTTCCACGTCCTCTTGCATAAATAAATAAACCGTTTTGAAACCAAACTCGCCTCTCTCTCTCCTTTTTGAAATATTTTTGGGTTTGTTTGATCCTTTCCTTCCCAATCTCTCTTGTTTAATATATATTCATTTATATCACGCTCTCTTTTTATCTTCCTTTTTTTCCTCTCTCTTGTATTCTTCCTTCCCCTTTCTACTCAAACCAAGAAGAAAAAGAAAAGGTCAATCTTTGTTAAAGAATAGGATCTTCTACTACATCAGCTTTTAGATTTTTCACGCTTACTGCTTTTTTCTTCCCAAGATCGAAAATTTACTGAATTAACAACTAGAATGGTTTCTGAATTGATTAAAGAAAACATGCATATGAAGTTGTATATGGAAGGTACTGTTAACAATCATCATTTCAAGTGTACATCTGAAGGTGAAGGTAAACCATATGAAGGTACTCAAACTATGAGAATTAAAGCTGTTGAAGGTGGTCCATTGCCATTTGCTTTTGATATTTTGGCTACTTCTTTCATGTATGGTTCTAAGACTTTCATTAACCATACTCAAGGTATTCCAGATTTTTTCAAGCAATCTTTTCCAGAAGGTTTTACTTGGGAAAGAGTTACTACTTACGAAGATGGTGGTGTTTTGACTGCTACTCAAGATACTTCTTTGCAAGATGGTTGTTTGATTTACAACGTTAAGATTAGAGGTGTTAATTTTCCATCTAATGGTCCAGTTATGCAAAAAAAGACTTTGGGTTGGGAAGCATCTACTGAAACTTTGTATCCAGCTGATGGTGGTTTGGAAGGTAGAGCTGATATGGCTTTGAAATTGGTTGGTGGTGGTCATTTGATTTGTAACTTGAAAACTACTTACAGATCAAAAAAGCCAGCTAAGAATTTGAAAATGCCAGGTGTTTATTACGTTGATAGAAGATTGGAAAGAATTAAAGAAGCTGATAAAGAAACTTACGTTGAACAACATGAAGTTGCTGTTGCAAGATATTGTGATTTGCCATCTAAATTGGGTCATAGATAAgaattcTTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTGaagcttATCGATACCGTCGACCTCGAGttCTTTTaGTTTCAGCTTTCCGCaacaGTATAATTTTATAAAcCCTGGttGGTACCCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGTCAGATATAGTCGGATTGCCCTTTTAAGCAATCAATAGTGTTTTATTTGCAACAATGTCGTCATAGTTTAATATGTCCTATAAGATGTTAACTTGCTCAACATTCAACAAAGTTTGGTCTCTTGGCCCAGTTGGTTAAGGCACCGTGCTAATAACGCGGGGATCAGCGGTTCGATCCCGCTAGAGACCATTTATTTTTAAATTCGACCGTCTGACAGACGGTGCACTTATTACTTAATTTTTTTTTTTTTTTAATCCTCGTTAAACTGAGTACATAATCAGCAACAATATAATATAAAAGTTTTGCTTTTTTATGTACGTACAATAGGGGTACTTAAATAAATAAATAAACAATAAAACAATCAATAAATAAATAATATACGATCATAAATTACGAGATATATTTCATAGTCTTGACTTCTCCAGCCGAACGTCATTAAATTTAAGGCAGAAAAGATAAACAAGGG")
pRG722 <- c(pRG722_up, pRG722_down)

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG722[1:50], subject = LSY4811_11B[["chrXII"]]))
end <- end(matchPattern(pattern = pRG722[(length(pRG722)-50):length(pRG722)], subject = LSY4811_11B[["chrXII"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG722) - (end - start)
LSY4811_11B[["chrXII"]][(start - nt_added + 1):end] <- pRG722

# is the sequence present now?
matchPattern(pattern = pRG722, subject = LSY4811_11B[["chrXII"]])

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4811_11B[["chrXII"]])

# mask endogenous sequences corresponding to reporter parts
P_ACT1 <- DNAString(x = "GAAGCGGGTAAGCTGCCACAGCAATTAATGCACAACATTTAACCTACATTCTTCCTTATCGGATCCTCAAAACCCTTAAAAACATATGCCTCACCCTAACATATTTTCCAATTAACCCTCAATATTTCTCTGTCACCCGGCCTCTATTTTCCATTTTCTTCTTTACCCGCCACGCGTTTTTTTCTTTCAAATTTTTTTCTTCCTTCTTCTTTTTCTTCCACGTCCTCTTGCATAAATAAATAAACCGTTTTGAAACCAAACTCGCCTCTCTCTCTCCTTTTTGAAATATTTTTGGGTTTGTTTGATCCTTTCCTTCCCAATCTCTCTTGTTTAATATATATTCATTTATATCACGCTCTCTTTTTATCTTCCTTTTTTTCCTCTCTCTTGTATTCTTCCTTCCCCTTTCTACTCAAACCAAGAAGAAAAAGAAAAGGTCAATCTTTGTTAAAGAATAGGATCTTCTACTACATCAGCTTTTAGATTTTTCACGCTTACTGCTTTTTTCTTCCCAAGATCGAAAATTTACTGAATTAACA")
matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4811_11B[["chrVI"]])
LSY4811_11B[["chrVI"]][matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4811_11B[["chrVI"]])] <- DNAString(x = paste0(rep("N", length(P_ACT1)), collapse = ""))
matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4811_11B[["chrVI"]])

T_CYC2 <- DNAString(x = "TTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTG")
matchPattern(pattern = T_CYC2, subject = LSY4811_11B[["chrXV"]])
LSY4811_11B[["chrXV"]][matchPattern(pattern = T_CYC2, subject = LSY4811_11B[["chrXV"]])] <- DNAString(x = paste0(rep("N", length(T_CYC2)), collapse = ""))
matchPattern(pattern = T_CYC2, subject = LSY4811_11B[["chrXV"]])

# -----------------------------------------------------------------------------
# export to fasta file
export(object = LSY4811_11B, con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Data/S_cerevisiae_reference_genomes/S288C_R64-2-1_W303_SNPs_MATa_hml_hmr_LSY4811-11B.fasta")

# =============================================================================
# LSY4812-5A modifications
LSY4812_5A <- genome

# -----------------------------------------------------------------------------
pRG745 <- DNAString(x = "CCTCCTCTAGTACACTCTATATTTTTTTATGCCTCGGTAATGATTTTCATTTTTTTTTTTCCACCTAGCGGATGACTCTTTTTTTTTCTTAGCGATTGGCATTATCACATAATGAATTATACATTATATAAAGTAATGTGATTTCTTCGAAGAATATACTAAAAAATGAGCAGGCAAGATAAACGAAGGCACATATGAGGAGGGCTTTTGTAGAAAGAAATACGAACGAAACGAAAATCAGCGTTGCCATCGCTTTGGACAAAGCTCCCTTACCTGAAGAGTCGAATTTTATTGATGAACTTATAACTTCCAAGCATGCAAACCAAAAGGGAGAACAAGTAATCCAAGTAGACACGGGAATTGGATTCTTGGATCACATGTATCATGCACTGGCTAAACATGCAGGCTGGAGCTTACGACTTTACTCAAGAGGTGATTTAATCATCGATGATCATCACACTGCAGAAGATACTGCTATTGCACTTGGTATTGCATTCAAGCAGGCTATGGGTAACTTTGCCGGTGTTAAAAGATTTGGACATGCTTATTGTCCACTTGACGAAGCTCTTTCCAGAAGCGTAGTTGACTTGTCGGGACGGCCCTATGCTGTTATCGATTTGGGATTAAAGCGTGAAAAGGTTGGGGAATTGTCCTGTGAAATGATCCCTCACTTACTATATTCCTTTTCGGTAGCTGCTGGAATTACTTTGCATGTTACCTGCTTATATGGTAGTAATGACCATCATCGTGCTGAAAGCGCTTTTAAATCTCTGGCTGTTGCCATGCGCGCGGCTACTTCTCTTACTGGAAGTTCTGAAGTCCCAAGCACGAAGGGAGTGTTGTAAAGATGAATTGGATTATGTCAGGAAAAGAACGACAATTTTGCATCCAAATTGTCTAAATTTTAGAGTTGCTTGAAAACAATAGAACCTTACTTGCTTTATAATTACGTTAATTAGAAGCGTTATCTCGTGAAGGAATATAGTACGTAGCCGTATAAATTGAATTGAATGTTCAGCTTATAGAATAGAGACACTTTGCTGTTCAATGCGTCGTCACTTACCATACTCACTTTATTATACGACTTTAAGTATAAACTCCGCGGTTATGGTAAAATTAATGATGCACAAACGTCCGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTGGAGCTCCACCGCGGTGGCGGCCGCTCTAGAGAAGCGGGTAAGCTGCCACAGCAATTAATGCACAACATTTAACCTACATTCTTCCTTATCGGATCCTCAAAACCCTTAAAAACATATGCCTCACCCTAACATATTTTCCAATTAACCCTCAATATTTCTCTGTCACCCGGCCTCTATTTTCCATTTTCTTCTTTACCCGCCACGCGTTTTTTTCTTTCAAATTTTTTTCTTCCTTCTTCTTTTTCTTCCACGTCCTCTTGCATAAATAAATAAACCGTTTTGAAACCAAACTCGCCTCTCTCTCTCCTTTTTGAAATATTTTTGGGTTTGTTTGATCCTTTCCTTCCCAATCTCTCTTGTTTAATATATATTCATTTATATCACGCTCTCTTTTTATCTTCCTTTTTTTCCTCTCTCTTGTATTCTTCCTTCCCCTTTCTACTCAAACCAAGAAGAAAAAGAAAAGGTCAATCTTTGTTAAAGAATAGGATCTTCTACTACATCAGCTTTTAGATTTTTCACGCTTACTGCTTTTTTCTTCCCAAGATCGAAAATTTACTGAATTAACAACTAGAATGAAAGGTGAAGAATTATTCACTGGTGTTGTCCCAATTTTGGTTGAATTAGATGGTGATGTTAATGGTCACAAATTTTCTGTCTCCGGTGAAGGTGAAGGTGATGCTACTTACGGTAAATTGACCTTAAAATTTATTTGTACTACTGGTAAATTGCCAGTTCCATGGCCAACCTTAGTCACTACTTTAGGTTATGGTTTGATGTGTTTTGCTAGATACCCAGATCATATGAAACAACATGACTTTTTCAAGTCTGCCATGCCAGAAGGTTATGTTCAAGAAAGAACTATTTTTTTCAAAGATGACGGTAACTACAAGACCAGAGCTGAAGTCAAGTTTGAAGGTGATACCTTAGTTAATAGAATCGAATTAAAAGGTATTGATTTTAAAGAAGATGGTAACATTTTAGGTCACAAATTGGAATACAACTATAACTCTCACAATGTTTACATCATGGCTGACAAACAAAAGAATGGTATCAAAGTTAACTTCAAAATTAGACACAACATTGAAGATGGTTCTGTTCAATTAGCTGACCATTATCAACAAAATACTCCAATTGGTGATGGTCCAGTCTTGTTACCAGACAACCATTACTTATCCTATCAATCTAAGTTATCCAAAGATCCAAACGAAAAGAGAGACCACATGGTCTTGTTAGAATTTGTTACTGCTGCTGGTATTACCCATGGTATGGATGAATTGTACAAAGaattcTTAGACAACACCCGCTTACGCATAGCTATTCAGAAATCAGGCCGTTTAAGCGATGATTCACGAGAATTGCTGGCCCGCTGCGGCATAAAAATTAATTTACACACTCAGCGCCTGATTaAGCTTttttCTTTTaGTTTCAGCTTTCCGCaacaGTATAATTTTATAAAcCCTGGtTACCCATACGATGTTCCAGATTACGCTtaaCTCGAGACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTTGgtaccCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGACACCGATTATTTAAAGCTGCAGCATACGATATATATACATGTGTATATATGTATACCTATGAATGTCAGTAAGTATGTATACGAACAGTATGATACTGAAGATGACAAGGTAATGCATCATTCTATACGTGTCATTCTGAACGAGG")

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG745[1:50], subject = LSY4812_5A[["chrXV"]]))
end <- end(matchPattern(pattern = pRG745[(length(pRG745)-50):length(pRG745)], subject = LSY4812_5A[["chrXV"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG745) - (end - start)
LSY4812_5A[["chrXV"]][(start - nt_added + 1):end] <- pRG745

# is the sequence present now?
matchPattern(pattern = pRG745, subject = LSY4812_5A[["chrXV"]])

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4812_5A[["chrXV"]])

# mask endogenous sequences corresponding to reporter parts
T_CYC1 <- DNAString(x = "ACAGGCCCCTTTTCCTTTGTCTATATCATGTAATTAGTTATGTCACGCTTACATTCACGCCCTCCTCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTTAATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAAACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGCAAGgTTCGCAGTTTACACTCTCATCGTCGCTCTCATCATCGCTTCCGTTGTTGTTTTCCTTAGTAGCGTCTGCTT")
matchPattern(pattern = T_CYC1, subject = LSY4812_5A[["chrX"]], max.mismatch = 2)
LSY4812_5A[["chrX"]][matchPattern(pattern = T_CYC1, subject = LSY4812_5A[["chrX"]], max.mismatch = 2)] <- DNAString(x = paste0(rep("N", length(T_CYC1)), collapse = ""))
matchPattern(pattern = T_CYC1, subject = LSY4812_5A[["chrX"]], max.mismatch = 2)

P_ACT1 <- DNAString(x = "GAAGCGGGTAAGCTGCCACAGCAATTAATGCACAACATTTAACCTACATTCTTCCTTATCGGATCCTCAAAACCCTTAAAAACATATGCCTCACCCTAACATATTTTCCAATTAACCCTCAATATTTCTCTGTCACCCGGCCTCTATTTTCCATTTTCTTCTTTACCCGCCACGCGTTTTTTTCTTTCAAATTTTTTTCTTCCTTCTTCTTTTTCTTCCACGTCCTCTTGCATAAATAAATAAACCGTTTTGAAACCAAACTCGCCTCTCTCTCTCCTTTTTGAAATATTTTTGGGTTTGTTTGATCCTTTCCTTCCCAATCTCTCTTGTTTAATATATATTCATTTATATCACGCTCTCTTTTTATCTTCCTTTTTTTCCTCTCTCTTGTATTCTTCCTTCCCCTTTCTACTCAAACCAAGAAGAAAAAGAAAAGGTCAATCTTTGTTAAAGAATAGGATCTTCTACTACATCAGCTTTTAGATTTTTCACGCTTACTGCTTTTTTCTTCCCAAGATCGAAAATTTACTGAATTAACA")
matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4812_5A[["chrVI"]])
LSY4812_5A[["chrVI"]][matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4812_5A[["chrVI"]])] <- DNAString(x = paste0(rep("N", length(P_ACT1)), collapse = ""))
matchPattern(pattern = reverseComplement(P_ACT1), subject = LSY4812_5A[["chrVI"]])

# -----------------------------------------------------------------------------
pRG738_up <- DNAString(x = "CCCTTAGCTCTCATTATTTTTTGCTTTTTCTCTTGAGGTCACATGATCGCAAAATGGCAAATGGCACGTGAAGCTGTCGATATTGGGGAACTGTGGTGGTTGGCAAATGACTAATTAAGTTAGTCAAGGCGCCATCCTCATGAAAACTGTGTAACATAATAACCGAAGTGTCGAAAAGGTGGCACCTTGTCCAATTGAACACGCTCGATGAAAAAAATAAGATATATATAAGGTTAAGTAAAGCGTCTGTTAGAAAGGAAGTTTTTCCTTTTTCTTGCTCTCTTGTCTTTTCATCTACTATTTCCTTCGTGTAATACAGGGTCGTCAGATACATAGATACAATTCTATTACCCCCATCCATACAATGCCATCTCATTTCGATACTGTTCAACTACACGCCGGTCAAGAGAACCCTGGTGACAATGCTCACAGATCCAGAGCTGTACCAATTTACGCCACCACTTCTTATGTTTTCGAAAACTCTAAGCATGGTTCGCAATTGTTTGGTTTAGAAGTTCCAGGTTACGTCTATTCCCGTTTCCAAAACCCAACCAGTAATGTTTTGGAAGAAAGAATTGCTGCTTTAGAAGGTGGTGCTGCTGCTTTGGCTGTTTCCTCCGGTCAAGCCGCTCAAACCCTTGCCATCCAAGGTTTGGCACACACTGGTGACAACATCGTTTCCACTTCTTACTTATACGGTGGTACTTATAACCAGTTCAAAATCTCGTTCAAAAGATTTGGTATCGAGGCTAGATTTGTTGAAGGTGACAATCCAGAAGAATTTGAAAAGGTCTTTGATGAAAGAACCAAGGCTGTTTATTTGGAAACCATTGGTAATCCAAAGTACAATGTTCCGGATTTTGAAAAAATTGTTGCAATTGCTCACAAACACGGTATTCCAGTTGTCGTTGACAACACATTTGGTGCCGGTGGTTACTTCTGTCAGCCAATTAAATACGGTGCTGATATTGTAACACATTCTGCTACCAAATGGATTGGTGGTCATGGTACTACTATCGGTGGTATTATTGTTGACTCTGGTAAGTTCCCATGGAAGGACTACCCAGAAAAGTTCCCTCAATTCTCTCAACCTGCCGAAGGTTATCACGGTACTATCTACAATGAAGCCTACGGTAACTTGGCATACATCGTTCATGTTAGAACTGAACTATTAAGAGATTTGGGTCCATTGATGAACCCATTTGCCTCTTTCTTGCTACTACAAGGTGTTGAAACATTATCTTTGAGAGCTGAAAGACACGGTGAAAATGCATTGAAGTTAGCCAAATGGTTAGAACAATCCCCATACGTATCTTGGGTTTCATACCCTGGTTTAGCATCTCATTCTCATCATGAAAATGCTAAGAAGTATCTATCTAACGGTTTCGGTGGTGTCTTATCTTTCGGTGTAAAAGACTTACCAAATGCCGACAAGGAAACTGACCCATTCAAACTTTCTGGTGCTCAAGTTGTTGACAATTTAAAGCTAGCCTCTAACTTGGCCAATGTTGGTGATGCCAAGACCTTAGTCATTGCTCCATACTTCACTACCCACAAACAATTAAATGACAAAGAAAAGTTGGCATCTGGTGTTACCAAGGACTTAATTCGTGTCTCTGTTGGTATCGAATTTATTGATGACATTATTGCAGACTTCCAGCAATCTTTTGAAACTGTTTTCGCTGGCCAAAAACCATGAGTGTGCGTAATGAGTTGTAAAATTATGTATAAACCTACTTTCTCTCACAAGTACTATACTTTTATAAAACGAACTTTATTGAAATGAATATCCTTTTTTTCCCTTGTTACATGTCGTGACTCGTACTTTGAACCTAAATTGTTCTAACATCAAAGAACAGTGTTAATTCGCAGTCGAGAAGAAAAATATGGTGAACAAGACTCATCTACTTCATGAGACTACTTTACGCCTCCTATAAAGCTGTCACACTGGATAAATTTATTGTAGGACCAAGTTACAAAAGAGGATGATGGAGGTTTCTTTACAATAAAGAAGCACATGTGTGTTAACGTTTTTAGTATTTGCTTGTTATGTAAATCAGGAAAACTTCGCGGGATTTGGTTGGATGCTACTTTCCATACAATAAATATTATAGATCAAAAAAGCCAAATTACAAGTAAAGATTAGTAAAGCTGTTGCAATTCCATCGTTGATAAAAATGTTAGTTATTAAATATAAAAGTCAGAATAGGTGAACTTGGATTTAATTGTTGGCATTTCGTTGCTGCTAGAGGCCATAATATTAGATAGCCAGGACATACTAGGTCTCCTCGTGGTATAGGAATCCATAAAATGGAATTGGTGACGATCGGATGTGCTGCAAGGCGATTAAGTTGGGTAACGCCAGGGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTGAATTGTAATACGACTCACTATAGGGCGAATTGGAGAAACAGAATAATATATAAACCTGTATAATATAACCTTGAAGACTATATTTCTTTTCGCGGCCGCCTATAGGGCGAATTG")
pRG738_down <- DNAString(x = "GAGCTCCACCGCGGTGGCTCTAGAATGGTTTCTGAATTGATTAAAGAAAACATGCATATGAAGTTGTATATGGAAGGTACTGTTAACAATCATCATTTCAAGTGTACATCTGAAGGTGAAGGTAAACCATATGAAGGTACTCAAACTATGAGAATTAAAGCTGTTGAAGGTGGTCCATTGCCATTTGCTTTTGATATTTTGGCTACTTCTTTCATGTATGGTTCTAAGACTTTCATTAACCATACTCAAGGTATTCCAGATTTTTTCAAGCAATCTTTTCCAGAAGGTTTTACTTGGGAAAGAGTTACTACTTACGAAGATGGTGGTGTTTTGACTGCTACTCAAGATACTTCTTTGCAAGATGGTTGTTTGATTTACAACGTTAAGATTAGAGGTGTTAATTTTCCATCTAATGGTCCAGTTATGCAAAAAAAGACTTTGGGTTGGGAAGCATCTACTGAAACTTTGTATCCAGCTGATGGTGGTTTGGAAGGTAGAGCTGATATGGCTTTGAAATTGGTTGGTGGTGGTCATTTGATTTGTAACTTGAAAACTACTTACAGATCAAAAAAGCCAGCTAAGAATTTGAAAATGCCAGGTGTTTATTACGTTGATAGAAGATTGGAAAGAATTAAAGAAGCTGATAAAGAAACTTACGTTGAACAACATGAAGTTGCTGTTGCAAGATATTGTGATTTGCCATCTAAATTGGGTCATAGATAAgaattcTTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTGaagcttAACCTGTCTGATATCACCGCATGCAATTAGTATCTTTTTTAGAAGGTTCCTCTTAAAATCTTTATCGGCAAATGCTGACCCGGTACCACCTCGAGttCTTTTaGTTTCAGCTTTCCGCaacaGTATAATTTTATAAAcCCTGGttGGTACCCAGCTTCGGCCGTATTTACAGATAACGATGTCGTTATTATATATATATATATATATTGTCACTTGTACCCAGCTTTTGTTCCCTTTAGTGAGGGTTAATTTCGAGCTTGGCGTAATCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAACAGCTGTCAGATATAGTCGGATTGCCCTTTTAAGCAATCAATAGTGTTTTATTTGCAACAATGTCGTCATAGTTTAATATGTCCTATAAGATGTTAACTTGCTCAACATTCAACAAAGTTTggtctcttggcccagttggttaaggcaccgtgctaataacgcggggatcagcggttcgatcccgctagagaccaTTTATTTTTAAATTCGACCGTCTGACAGACGGTGCACTTATTACTTAATTTTTTTTTTTTTTTAATCCTCGTTAAACTGAGTACATAATCAGCAACAATATAATATAAAAGTTTTGCTTTTTTATGTACGTACAATAGGGGTACTTAAATAAATAAATAAACAATAAAACAATCAATAAATAAATAATATACGATCATAAATTACGAGATATATTTCATAGTCTTGACTTCTCCAGCCGAACGTCATTAAATTTAAGGCAGAAAAGATAAACAAGGG")
pRG738 <- c(pRG738_up, pRG738_down)

# find targeting sequences in genome
start <- start(matchPattern(pattern = pRG738[1:50], subject = LSY4812_5A[["chrXII"]]))
end <- end(matchPattern(pattern = pRG738[(length(pRG738)-50):length(pRG738)], subject = LSY4812_5A[["chrXII"]]))

# let's substitute sequence upstream of HIS3 to accomodate the insertion
nt_added <- length(pRG738) - (end - start)
LSY4812_5A[["chrXII"]][(start - nt_added + 1):end] <- pRG738

# is the sequence present now?
matchPattern(pattern = pRG738, subject = LSY4812_5A[["chrXII"]])

# HOcs insertion?
HOcs <- DNAString(x = "TTCTTTTAGTTTCAGCTTTCCGCAACAGTATAATTTTATAAACCCTGGTT")
matchPattern(pattern = HOcs, subject = LSY4812_5A[["chrXII"]])

# mask endogenous sequences corresponding to reporter parts
T_CYC2 <- DNAString(x = "TTTAACGTTCGAGATTAAGAAGGCAATTACAATAATATTACATACTGTAGATAATGAATATAATGAAAAATATAAGAGTTTAAACTATACATTGTTAAAGCCAAACTAAGATATTATATTCATTTCCTTGGCATAACGTGTAGTAACTCTTTGaagcttAACCTGTCTGATATCACCGCATGCAATTAGTATCTTTTTTAGAAGGTTCCTCTTAAAATCTTTATCGGCAAATGCTGACCCGGTACCAC")
matchPattern(pattern = T_CYC2, subject = LSY4812_5A[["chrXV"]], with.indels = TRUE, max.mismatch = 6)
LSY4812_5A[["chrXV"]][matchPattern(pattern = T_CYC2, subject = LSY4812_5A[["chrXV"]], with.indels = TRUE, max.mismatch = 6)] <- DNAString(x = paste0(rep("N", 242), collapse = ""))  # there is an inserted HindIII site (6 nt)
matchPattern(pattern = T_CYC2, subject = LSY4812_5A[["chrXV"]], with.indels = TRUE, max.mismatch = 6)

# -----------------------------------------------------------------------------
# export to fasta file
export(object = LSY4812_5A, con = "/home/robert/Research/Manuscripts/My_manuscripts/20-04-17-MRX_nicking_manuscript/Data/S_cerevisiae_reference_genomes/S288C_R64-2-1_W303_SNPs_MATa_hml_hmr_LSY4812-5A.fasta")
